<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>3. Configuration &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="4. Tutorial (Command line)" href="tutorial_cli.html" />
    <link rel="prev" title="2. Overview" href="overview.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">2. Overview</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">3. Configuration</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#config-file">3.1. config file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#alias-lookups">3.2. Alias lookups</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#level-lookup">3.2.1. 1-level lookup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">3.2.2. 2-level lookup</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#parameters">3.2.2.1. Parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alternative-parameters">3.2.2.2. Alternative parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#level-lookup-file-formats">3.2.3. 2-level lookup file formats</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#cluster">3.2.3.1. cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome">3.2.3.2. genome</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reference">3.2.3.3. reference</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_cli.html">4. Tutorial (Command line)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_py.html">5. Tutorial (Python)</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">6. Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">7. Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">8. Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">9. Examples</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">3. </span>Configuration</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/configuration.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="configuration">
<h1><span class="section-number">3. </span>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline"></a></h1>
<p>Note that all command line code snippets in the following tutorial are for <strong>bash terminal</strong>. You may have to adapt them according to your operating system.</p>
<section id="config-file">
<h2><span class="section-number">3.1. </span>config file<a class="headerlink" href="#config-file" title="Permalink to this headline"></a></h2>
<p>The config.ini file is used for two main purposes: setting default values and assigning short aliases to long values. All entries in the config.ini file that end with ‘alias’ or ‘sets’ are used to assign aliases, while the rest are for setting default values. Use the following to export the config file location as an environment variable:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">export</span> <span class="nv">MINORG_CONFIG</span><span class="o">=</span>/path/to/config.ini
</pre></div>
</div>
<p>If you’d like the config file to apply to all users, consider setting it globally.</p>
</section>
<section id="alias-lookups">
<h2><span class="section-number">3.2. </span>Alias lookups<a class="headerlink" href="#alias-lookups" title="Permalink to this headline"></a></h2>
<p>There are two types of alias lookups in MINORg: 1-level lookups and 2-level lookups.</p>
<section id="level-lookup">
<h3><span class="section-number">3.2.1. </span>1-level lookup<a class="headerlink" href="#level-lookup" title="Permalink to this headline"></a></h3>
<p>1-level lookups are defined directly in the config file. Parameters that use 1-level lookups are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--assembly</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--annotation</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--db</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--domain</span></code></p></li>
</ul>
<p>To understand how aliases are assigned using the config file, let us take a look at <code class="docutils literal notranslate"><span class="pre">assembly</span> <span class="pre">alias</span></code> under section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>. Here, we assign aliases for FASTA files of reference assemblies in the format &lt;semicolon-separated alias&gt;:&lt;value&gt;.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">assembly</span> <span class="n">alias</span> <span class="o">=</span> <span class="n">tair10</span><span class="p">;</span><span class="n">TAIR10</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">fasta</span>
                 <span class="n">araport11</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">fasta</span>
                 <span class="n">araly2</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araly2</span><span class="o">.</span><span class="n">fasta</span>
                 <span class="n">araha1</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araha1</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>With this setup, we can use <code class="docutils literal notranslate"><span class="pre">--assembly</span> <span class="pre">TAIR10</span></code> instead of <code class="docutils literal notranslate"><span class="pre">--assembly</span> <span class="pre">/path/to/subset_ref_TAIR10.fasta</span></code> when building our MINORg command (although either way is acceptable). This applies to <code class="docutils literal notranslate"><span class="pre">annotation</span> <span class="pre">alias</span></code> (<code class="docutils literal notranslate"><span class="pre">--annotation</span></code>), <code class="docutils literal notranslate"><span class="pre">rps</span> <span class="pre">database</span> <span class="pre">alias</span></code> (<code class="docutils literal notranslate"><span class="pre">--db</span></code>), and <code class="docutils literal notranslate"><span class="pre">gff</span> <span class="pre">attribute</span> <span class="pre">modification</span> <span class="pre">presets</span></code> (<code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code>) as well.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%" />
<col style="width: 13%" />
<col style="width: 15%" />
<col style="width: 15%" />
<col style="width: 42%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>assembly</strong></p></td>
<td><p><strong>annotation</strong></p></td>
<td><p><strong>rps database</strong></p></td>
<td><p><strong>attr mod</strong></p></td>
<td><p><strong>description</strong></p></td>
</tr>
<tr class="row-even"><td><p>assembly alias</p></td>
<td><p>annotation
alias</p></td>
<td><p>rps database
alias</p></td>
<td><p>gff attribute
modification
presets</p></td>
<td><p>[config] assign alias to values</p></td>
</tr>
<tr class="row-odd"><td><p>--assembly</p></td>
<td><p>--annotation</p></td>
<td><p>--db</p></td>
<td><p>--attr-mod</p></td>
<td><p>[CLI] specify value (alias or raw)</p></td>
</tr>
</tbody>
</table>
<p>Unlike the above parameters, <code class="docutils literal notranslate"><span class="pre">domain</span> <span class="pre">alias</span></code> is specified in the reverse format using commas instead of semi-colons (&lt;value&gt;:&lt;comma-separated aliases) for readability reasons, as most PSSM-Ids are comprised of the same number of digits. Therefore, the <code class="docutils literal notranslate"><span class="pre">domain</span> <span class="pre">alias</span></code> in section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code> looks like this instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">domain</span> <span class="n">alias</span> <span class="o">=</span> <span class="mi">366714</span><span class="p">:</span><span class="n">TIR</span>
               <span class="mi">366375</span><span class="p">:</span><span class="n">NB</span><span class="o">-</span><span class="n">ARC</span><span class="p">,</span><span class="n">NBS</span>
               <span class="mi">375519</span><span class="p">:</span><span class="n">Rx_N</span>
</pre></div>
</div>
<p>As with the above parameters, you can use <code class="docutils literal notranslate"><span class="pre">--domain</span> <span class="pre">TIR</span></code> instead of <code class="docutils literal notranslate"><span class="pre">--domain</span> <span class="pre">366714</span></code> (although either way is acceptable). Do note, however, that if you use <code class="docutils literal notranslate"><span class="pre">--domain</span> <span class="pre">TIR</span></code>, newly generate file and sequence names will use <code class="docutils literal notranslate"><span class="pre">TIR</span></code> instead of <code class="docutils literal notranslate"><span class="pre">366714</span></code> if the domain is to be included in the name.</p>
</section>
<section id="id1">
<h3><span class="section-number">3.2.2. </span>2-level lookup<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h3>
<p>2-level lookups are defined using a file that maps aliases to values. This means that there is a first layer of lookup that maps aliases to those files, and then a second layer of lookup in those files that maps aliases to values. Parameters that use 2-level lookups are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--reference</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--cluster</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--indv</span></code></p></li>
</ul>
<p>To understand how aliases are assigned, we shall use the sample_config.ini file (provided at <a class="reference external" href="https://github.com/rlrq/MINORg/blob/master/examples/sample_config.ini">https://github.com/rlrq/MINORg/blob/master/examples/sample_config.ini</a>) and zoom in on reference genomes as an example. First, let use look at <code class="docutils literal notranslate"><span class="pre">reference</span> <span class="pre">sets</span></code> under the section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>. Here, we assign aliases for lookup files (which at the command line we can specify using <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="n">sets</span> <span class="o">=</span> <span class="n">athaliana</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">athaliana_genomes</span><span class="o">.</span><span class="n">txt</span>
                 <span class="n">arabidopsis</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">arabidopsis_genomes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Each lookup file contains mapping of aliases to reference genome files as well as meta data. Let us look at the file athaliana_genomes.txt (provided at <a class="reference external" href="https://github.com/rlrq/MINORg/blob/master/examples/athaliana_genomes.txt">https://github.com/rlrq/MINORg/blob/master/examples/athaliana_genomes.txt</a>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tair10</span><span class="p">;</span><span class="n">TAIR10</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">fasta</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">gff</span>  <span class="mi">1</span>
<span class="n">araport11</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">fasta</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">gff</span>       <span class="mi">1</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: path to genome FASTA file</p></li>
<li><p>column 3: path to genome GFF3 file</p></li>
<li><p>column 4: NCBI genetic code number or name (optional if standard genetic code)</p></li>
<li><p>column 5: mapping of nonstandard GFF3 attribute field names to standard field names (optional if standard)</p></li>
</ul>
<p>By using <code class="docutils literal notranslate"><span class="pre">--reference-set</span> <span class="pre">athaliana</span></code> in the command line execution, we can use <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">TAIR10</span></code> (or <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">tair10</span></code>) to tell MINORg to use ‘/path/to/subset_ref_TAIR10.fasta’ and ‘/path/to/subset_ref_TAIR10.gff’ as reference assembly and annotation files respectively without having to type their paths out. In fact, we can even specify multiple reference genomes using <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">TAIR10,araport11</span></code> and making sure to use a comma to separate reference genome aliases.</p>
<p>But there’s more! We can set ‘athaliana’ as the default reference set AND ‘TAIR10’ as the default reference in the <code class="docutils literal notranslate"><span class="pre">[data]</span></code> section of the config file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="o">=</span> <span class="n">TAIR10</span>
<span class="n">reference</span> <span class="nb">set</span> <span class="o">=</span> <span class="n">athaliana</span>
</pre></div>
</div>
<p>This way, unless you wish to use a different reference genome from the default, you won’t have to type <code class="docutils literal notranslate"><span class="pre">--reference-set</span> <span class="pre">athaliana</span> <span class="pre">--reference</span> <span class="pre">TAIR10</span></code> either! This is particularly useful if you primarily design gRNA for only a single species. Of course, if you wish, you can combine all reference set files into a single massive file containing the mapping information for all possible reference genomes instead of having multiple files. However, I personally find it easier to maintain smaller files with descriptive names.</p>
<p>You can view the reference genomes in the default reference set using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --references <span class="c1">## prints contents of default reference set file</span>

Valid genome aliases <span class="o">(</span>defined <span class="k">in</span> /path/to/athaliana_genomes.txt<span class="o">)</span>:

&lt;semicolon-separated genome alias<span class="o">(</span>es<span class="o">)</span>&gt;       &lt;FASTA file&gt;    &lt;GFF3 file&gt;     &lt;NCBI genetic code&gt;     &lt;attribute name mapping&gt;
tair10<span class="p">;</span>TAIR10        /path/to/subset_ref_Araport11.fasta     /path/to/subset_ref_Araport11.gff
araport11    /path/to/subset_ref_TAIR10.fasta        /path/to/subset_ref_TAIR10.gff
</pre></div>
</div>
<p>To view the reference genomes in a non-default reference set, use:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --reference-set arabidopsis --references <span class="c1">## prints contents of arabidopsis_genomes.txt instead</span>

Valid genome aliases <span class="o">(</span>defined <span class="k">in</span> /path/to/arabidopsis_genomes.txt<span class="o">)</span>:

&lt;semicolon-separated genome alias<span class="o">(</span>es<span class="o">)</span>&gt;       &lt;FASTA file&gt;    &lt;GFF3 file&gt;     &lt;NCBI genetic code&gt;     &lt;attribute name mapping&gt;
tair10<span class="p">;</span>TAIR10        /path/to/subset_ref_Araport11.fasta     /path/to/subset_ref_Araport11.gff
araport11    /path/to/subset_ref_TAIR10.fasta        /path/to/subset_ref_TAIR10.gff
araly2<span class="p">;</span>alyrata2      /path/to/subset_ref_Araly2.fasta        /path/to/subset_ref_Araly2.gff
araha1<span class="p">;</span>ahalleri1     /path/to/subset_ref_Araha1.fasta        /path/to/subset_ref_Araha1.gff
</pre></div>
</div>
<p>Note that you can also provide an alias mapping file that is not in the config file to MINORg by specifying the path to the file instead of using a non-existent alias (e.g. <code class="docutils literal notranslate"><span class="pre">--reference-set</span> <span class="pre">/path/to/arabidopsis_genomes.txt</span></code>).</p>
<section id="parameters">
<h4><span class="section-number">3.2.2.1. </span>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline"></a></h4>
<p>The same logic applies as well to <code class="docutils literal notranslate"><span class="pre">cluster</span> <span class="pre">sets</span></code>-<code class="docutils literal notranslate"><span class="pre">cluster</span> <span class="pre">set</span></code> (<code class="docutils literal notranslate"><span class="pre">--cluster-set</span></code>-<code class="docutils literal notranslate"><span class="pre">--cluster</span></code>-<code class="docutils literal notranslate"><span class="pre">--clusters</span></code>) and <code class="docutils literal notranslate"><span class="pre">genome</span> <span class="pre">sets</span></code>-<code class="docutils literal notranslate"><span class="pre">genome</span> <span class="pre">set</span></code> (<code class="docutils literal notranslate"><span class="pre">--genome-set</span></code>-<code class="docutils literal notranslate"><span class="pre">--indv</span></code>-<code class="docutils literal notranslate"><span class="pre">--genomes</span></code>), with the caveat that there is no option to set default clusters or query genomes.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 18%" />
<col style="width: 16%" />
<col style="width: 15%" />
<col style="width: 51%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>reference</strong></p></td>
<td><p><strong>cluster</strong></p></td>
<td><p><strong>genome</strong></p></td>
<td><p><strong>description</strong></p></td>
</tr>
<tr class="row-even"><td><p>reference sets</p></td>
<td><p>cluster sets</p></td>
<td><p>genome sets</p></td>
<td><p>[config] assign alias to lookup files</p></td>
</tr>
<tr class="row-odd"><td><p>reference set</p></td>
<td><p>cluster set</p></td>
<td><p>genome set</p></td>
<td><p>[config] set default lookup file</p></td>
</tr>
<tr class="row-even"><td><p>reference</p></td>
<td></td>
<td></td>
<td><p>[config] set default value</p></td>
</tr>
<tr class="row-odd"><td><p>--reference-set</p></td>
<td><p>--cluster-set</p></td>
<td><p>--genome-set</p></td>
<td><p>[CLI] specify lookup file (alias or path)</p></td>
</tr>
<tr class="row-even"><td><p>--reference</p></td>
<td><p>--cluster</p></td>
<td><p>--indv</p></td>
<td><p>[CLI] specify reference/cluster/indv
(comma-separated alias(es))</p></td>
</tr>
<tr class="row-odd"><td><p>--references</p></td>
<td><p>--clusters</p></td>
<td><p>--genomes</p></td>
<td><p>[CLI] print contents of lookup file to screen</p></td>
</tr>
</tbody>
</table>
</section>
<section id="alternative-parameters">
<h4><span class="section-number">3.2.2.2. </span>Alternative parameters<a class="headerlink" href="#alternative-parameters" title="Permalink to this headline"></a></h4>
<p>Do note that, because of the nature of these lookups, you cannot simply provide the value(s) mapped to an alias to <code class="docutils literal notranslate"><span class="pre">--reference</span></code>, <code class="docutils literal notranslate"><span class="pre">--cluster</span></code>, or <code class="docutils literal notranslate"><span class="pre">--indv</span></code>. If the desired files/genes are not specified in any mapping file, you will have to use the following alternatives:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 14%" />
<col style="width: 22%" />
<col style="width: 63%" />
</colgroup>
<tbody>
<tr class="row-odd"><td></td>
<td><p><strong>using aliases</strong></p></td>
<td><p><strong>not using aliases</strong></p></td>
</tr>
<tr class="row-even"><td><p><strong>reference</strong></p></td>
<td><p>--reference
&lt;alias(es)&gt;</p></td>
<td><p>--assembly &lt;path to FASTA&gt; --annotation &lt;path to GFF3&gt;
--genetic-code &lt;number or name&gt; --attr-mod &lt;attrinute
modiication&gt;*</p></td>
</tr>
<tr class="row-odd"><td><p><strong>cluster</strong></p></td>
<td><p>--cluster
&lt;alias(es)&gt;</p></td>
<td><p>--gene &lt;comma-separated gene IDs&gt;**</p></td>
</tr>
<tr class="row-even"><td><p><strong>indv</strong></p></td>
<td><p>--indv &lt;alias(es)&gt;</p></td>
<td><p>--query &lt;path to FASTA&gt;***</p></td>
</tr>
</tbody>
</table>
<p>* <code class="docutils literal notranslate"><span class="pre">--genetic-code</span></code> and <code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code> are optional if the reference genome uses the standard genetic code and standard GFF attribute field names respectively. Do note that you CANNOT SPECIFY MULTIPLE reference genomes if not using aliases.</p>
<p>** If not using aliases, each cluster must be processed separately (i.e. a different MINORg execution for each cluster), as MINORg has no way of knowing which gene belongs to which cluster if you use <code class="docutils literal notranslate"><span class="pre">--gene</span></code>.</p>
<p>*** You may specify multiple query FASTA files by using <code class="docutils literal notranslate"><span class="pre">--query</span> <span class="pre">&lt;FASTA&gt;</span></code> as many times as needed (e.g. <code class="docutils literal notranslate"><span class="pre">--query</span> <span class="pre">&lt;FASTA</span> <span class="pre">1&gt;</span> <span class="pre">--query</span> <span class="pre">&lt;FASTA</span> <span class="pre">2&gt;</span> <span class="pre">--query</span> <span class="pre">&lt;FASTA</span> <span class="pre">3&gt;</span></code>).</p>
</section>
</section>
<section id="level-lookup-file-formats">
<h3><span class="section-number">3.2.3. </span>2-level lookup file formats<a class="headerlink" href="#level-lookup-file-formats" title="Permalink to this headline"></a></h3>
<section id="cluster">
<h4><span class="section-number">3.2.3.1. </span>cluster<a class="headerlink" href="#cluster" title="Permalink to this headline"></a></h4>
<p>The lookup files specified in <code class="docutils literal notranslate"><span class="pre">cluster</span> <span class="pre">sets</span></code> should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">RPS4</span><span class="p">;</span><span class="n">TTR1</span><span class="p">;</span><span class="n">RPS4_TTR1</span>   <span class="n">AT5G45050</span><span class="p">,</span><span class="n">AT5G45060</span><span class="p">,</span><span class="n">AT5G45200</span><span class="p">,</span><span class="n">AT5G45210</span><span class="p">,</span><span class="n">AT5G45220</span><span class="p">,</span><span class="n">AT5G45230</span><span class="p">,</span><span class="n">AT5G45240</span><span class="p">,</span><span class="n">AT5G45250</span>
<span class="n">RPS6</span>  <span class="n">AT5G46260</span><span class="p">,</span><span class="n">AT5G46270</span><span class="p">,</span><span class="n">AT5G46450</span><span class="p">,</span><span class="n">AT5G46470</span><span class="p">,</span><span class="n">AT5G46490</span><span class="p">,</span><span class="n">AT5G46510</span><span class="p">,</span><span class="n">AT5G46520</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: comma-separated gene IDs of genes in the cluster</p></li>
</ul>
</section>
<section id="genome">
<h4><span class="section-number">3.2.3.2. </span>genome<a class="headerlink" href="#genome" title="Permalink to this headline"></a></h4>
<p>The lookup files specified in <code class="docutils literal notranslate"><span class="pre">genome</span> <span class="pre">sets</span></code> should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">9654</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9654</span><span class="o">.</span><span class="n">fasta</span>
<span class="mi">9655</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9655</span><span class="o">.</span><span class="n">fasta</span>
<span class="mi">9944</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9944</span><span class="o">.</span><span class="n">fasta</span>
<span class="mi">9947</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9947</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: path to FASTA file</p></li>
</ul>
</section>
<section id="reference">
<h4><span class="section-number">3.2.3.3. </span>reference<a class="headerlink" href="#reference" title="Permalink to this headline"></a></h4>
<p>For ease of reference, the format for reference lookup files specified in <code class="docutils literal notranslate"><span class="pre">reference</span> <span class="pre">sets</span></code> is reproduced here:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tair10</span><span class="p">;</span><span class="n">TAIR10</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">fasta</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">gff</span>  <span class="mi">1</span>
<span class="n">araport11</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">fasta</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">gff</span>       <span class="mi">1</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: path to genome FASTA file</p></li>
<li><p>column 3: path to genome GFF3 file</p></li>
<li><p>column 4: NCBI genetic code number or name (optional if standard genetic code)</p></li>
<li><p>column 5: mapping of nonstandard GFF3 attribute field names to standard field names (optional if standard) (see <a class="reference internal" href="parameters.html#attribute-modification-format-reference-lookup-file"><span class="std std-ref">Attribute modification format (reference lookup file)</span></a> for format)</p></li>
</ul>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="overview.html" class="btn btn-neutral float-left" title="2. Overview" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial_cli.html" class="btn btn-neutral float-right" title="4. Tutorial (Command line)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>