<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>7. Parameters &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="8. Output" href="output.html" />
    <link rel="prev" title="6. Tutorial (Python)" href="tutorial_py.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">2. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">3. Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_docker.html">4. Tutorial (Docker)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_cli.html">5. Tutorial (Command line)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_py.html">6. Tutorial (Python)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">7. Parameters</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#cli-vs-python">7.1. CLI vs Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameter-types">7.2. Parameter types</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#flags-and-arguments">7.2.1. Flags and arguments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#flag">7.2.1.1. Flag</a></li>
<li class="toctree-l4"><a class="reference internal" href="#argument">7.2.1.2. Argument</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#paths">7.2.2. Paths</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#executables">7.2.2.1. Executables</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alias-lookup">7.2.3. Alias lookup</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#level-lookup">7.2.3.1. 1-level lookup</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id1">7.2.3.2. 2-level lookup</a></li>
<li class="toctree-l4"><a class="reference internal" href="#predefined-lookup">7.2.3.3. Predefined lookup</a></li>
<li class="toctree-l4"><a class="reference internal" href="#raw-values">7.2.3.4. Raw values</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#multiple-values">7.2.4. Multiple values</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#comma-separated-cli">7.2.4.1. Comma-separated (CLI)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multi-argument-cli">7.2.4.2. Multi-argument (CLI)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multi-value-list-python">7.2.4.3. Multi-value list (Python)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multi-value-dictionary-python">7.2.4.4. Multi-value dictionary (Python)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#parameter-descriptions">7.3. Parameter descriptions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#reference">7.3.1. Reference</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#reference-cli">7.3.1.1. Reference (CLI)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reference-python">7.3.1.2. Reference (Python)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#attribute-modification">7.3.2. Attribute modification</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#attribute-modification-format-cli">7.3.2.1. Attribute modification format (CLI)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#attribute-modification-format-reference-lookup-file">7.3.2.2. Attribute modification format (reference lookup file)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#attribute-modification-format-python">7.3.2.3. Attribute modification format (Python)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#extended-genome">7.3.3. Extended genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pam">7.3.4. PAM</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#ambiguous-bases-and-repeats">7.3.4.1. Ambiguous bases and repeats</a></li>
<li class="toctree-l4"><a class="reference internal" href="#spacers-and-3-or-5-pam">7.3.4.2. Spacers and 3’ or 5’ PAM</a></li>
<li class="toctree-l4"><a class="reference internal" href="#preset-pam-patterns">7.3.4.3. Preset PAM patterns</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#off-target-pattern">7.3.5. Off-target pattern</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#basic-unit">7.3.5.1. Basic unit</a></li>
<li class="toctree-l4"><a class="reference internal" href="#operators">7.3.5.2. Operators</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#prioritise-non-redundancy">7.3.6. Prioritise non-redundancy</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rps-blast-local-database">7.3.7. RPS-BLAST local database</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rps-blast-remote-database">7.3.8. RPS-BLAST remote database</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">8. Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">9. Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">10. Examples</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">7. </span>Parameters</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/parameters.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="parameters">
<h1><span class="section-number">7. </span>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline"></a></h1>
<p>In addition to a brief overview of the equivalent parameters between CLI and Python versions of MINORg, this section provides additional information about the input format and utility of some of the more cryptic parameters.</p>
<div class="contents local topic" id="contents">
<p class="topic-title">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#cli-vs-python" id="id3">CLI vs Python</a></p></li>
<li><p><a class="reference internal" href="#parameter-types" id="id4">Parameter types</a></p>
<ul>
<li><p><a class="reference internal" href="#flags-and-arguments" id="id5">Flags and arguments</a></p>
<ul>
<li><p><a class="reference internal" href="#flag" id="id6">Flag</a></p></li>
<li><p><a class="reference internal" href="#argument" id="id7">Argument</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#paths" id="id8">Paths</a></p>
<ul>
<li><p><a class="reference internal" href="#executables" id="id9">Executables</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#alias-lookup" id="id10">Alias lookup</a></p>
<ul>
<li><p><a class="reference internal" href="#level-lookup" id="id11">1-level lookup</a></p></li>
<li><p><a class="reference internal" href="#id1" id="id12">2-level lookup</a></p></li>
<li><p><a class="reference internal" href="#predefined-lookup" id="id13">Predefined lookup</a></p></li>
<li><p><a class="reference internal" href="#raw-values" id="id14">Raw values</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#multiple-values" id="id15">Multiple values</a></p>
<ul>
<li><p><a class="reference internal" href="#comma-separated-cli" id="id16">Comma-separated (CLI)</a></p></li>
<li><p><a class="reference internal" href="#multi-argument-cli" id="id17">Multi-argument (CLI)</a></p></li>
<li><p><a class="reference internal" href="#multi-value-list-python" id="id18">Multi-value list (Python)</a></p></li>
<li><p><a class="reference internal" href="#multi-value-dictionary-python" id="id19">Multi-value dictionary (Python)</a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#parameter-descriptions" id="id20">Parameter descriptions</a></p>
<ul>
<li><p><a class="reference internal" href="#reference" id="id21">Reference</a></p>
<ul>
<li><p><a class="reference internal" href="#reference-cli" id="id22">Reference (CLI)</a></p></li>
<li><p><a class="reference internal" href="#reference-python" id="id23">Reference (Python)</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#attribute-modification" id="id24">Attribute modification</a></p>
<ul>
<li><p><a class="reference internal" href="#attribute-modification-format-cli" id="id25">Attribute modification format (CLI)</a></p></li>
<li><p><a class="reference internal" href="#attribute-modification-format-reference-lookup-file" id="id26">Attribute modification format (reference lookup file)</a></p></li>
<li><p><a class="reference internal" href="#attribute-modification-format-python" id="id27">Attribute modification format (Python)</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#extended-genome" id="id28">Extended genome</a></p></li>
<li><p><a class="reference internal" href="#pam" id="id29">PAM</a></p>
<ul>
<li><p><a class="reference internal" href="#ambiguous-bases-and-repeats" id="id30">Ambiguous bases and repeats</a></p></li>
<li><p><a class="reference internal" href="#spacers-and-3-or-5-pam" id="id31">Spacers and 3’ or 5’ PAM</a></p></li>
<li><p><a class="reference internal" href="#preset-pam-patterns" id="id32">Preset PAM patterns</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#off-target-pattern" id="id33">Off-target pattern</a></p>
<ul>
<li><p><a class="reference internal" href="#basic-unit" id="id34">Basic unit</a></p></li>
<li><p><a class="reference internal" href="#operators" id="id35">Operators</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#prioritise-non-redundancy" id="id36">Prioritise non-redundancy</a></p></li>
<li><p><a class="reference internal" href="#rps-blast-local-database" id="id37">RPS-BLAST local database</a></p></li>
<li><p><a class="reference internal" href="#rps-blast-remote-database" id="id38">RPS-BLAST remote database</a></p></li>
</ul>
</li>
</ul>
</div>
<section id="cli-vs-python">
<h2><a class="toc-backref" href="#id3"><span class="section-number">7.1. </span>CLI vs Python</a><a class="headerlink" href="#cli-vs-python" title="Permalink to this headline"></a></h2>
<p>Several CLI arguments have no equivalents in the Python module as they were intended to simplify the building of commands for users who have little to no experience with coding. Users of the Python package are assumed to be comfortable with generating their own preset parameter combinations.</p>
<p>The table below lists the major similarities and differences between CLI arguments and the Python package’s MINORg class attributes (note that some attributes are in fact properties, but with the exception of <code class="docutils literal notranslate"><span class="pre">reference</span></code>, setting them should be no different from setting attributes).</p>
<p>Note: Some parameters only apply to specific subcommands (in addition to, of course, the full programme). The relevant subcommands will be indicated in square brackets in the ‘Category’ column.</p>
<p>Python attributes in the table below indicated with an asterisk (*) should be set using a dedicated method.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 16%" />
<col style="width: 29%" />
<col style="width: 29%" />
<col style="width: 27%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>Category</strong></p></td>
<td><p><strong>CLI arguments</strong></p></td>
<td><p><strong>Python attributes</strong></p></td>
<td><p><strong>Description</strong></p></td>
</tr>
<tr class="row-even"><td rowspan="3"><p>General</p></td>
<td><p>--directory</p></td>
<td><p>directory</p></td>
<td><p>output directory</p></td>
</tr>
<tr class="row-odd"><td><p>--prefix</p></td>
<td><p>prefix</p></td>
<td><p>output file/directory
prefix</p></td>
</tr>
<tr class="row-even"><td><p>--thread</p></td>
<td><p>thread</p></td>
<td><p>threads</p></td>
</tr>
<tr class="row-odd"><td rowspan="4"><p>Executable</p>
<p>(path to
executable</p>
<p>if not in
command-search</p>
<p>path)</p>
</td>
<td><p>--blastn</p></td>
<td><p>blastn</p></td>
<td><p>local BLAST’s blastn</p></td>
</tr>
<tr class="row-even"><td><p>--rpsblast</p></td>
<td><p>rpsblast</p></td>
<td><p>local BLAST’s
rpsblast/rpsblast+</p></td>
</tr>
<tr class="row-odd"><td><p>--mafft</p></td>
<td><p>mafft</p></td>
<td><p>MAFFT</p></td>
</tr>
<tr class="row-even"><td><p>--bedtools</p></td>
<td><p>bedtools</p></td>
<td><p>EXCEPTION: path to
directory containing
BEDTools executables</p></td>
</tr>
<tr class="row-odd"><td rowspan="7"><p>Reference
genomes</p>
<p>(CLI: seq,
full;</p>
<p>Python: seq,
filter)</p>
</td>
<td><p>--<a class="reference internal" href="#reference"><span class="std std-ref">reference</span></a></p></td>
<td><p><a class="reference internal" href="#reference"><span class="std std-ref">reference</span></a>*</p></td>
<td><p>reference genome</p></td>
</tr>
<tr class="row-even"><td><p>--assembly</p></td>
<td></td>
<td><p>reference genome FASTA</p></td>
</tr>
<tr class="row-odd"><td><p>--annotation</p></td>
<td></td>
<td><p>reference genome GFF</p></td>
</tr>
<tr class="row-even"><td><p>--<a class="reference internal" href="#attribute-modification"><span class="std std-ref">attr_mod</span></a></p></td>
<td></td>
<td><p>mapping for non-standard
GFF attribute field names</p></td>
</tr>
<tr class="row-odd"><td><p>--genetic-code</p></td>
<td></td>
<td><p>NCBI genetic code number
or name</p></td>
</tr>
<tr class="row-even"><td><p>--extend-gene</p></td>
<td></td>
<td><p>FASTA file of genes to
add to reference genome</p></td>
</tr>
<tr class="row-odd"><td><p>--extend-cds</p></td>
<td></td>
<td><p>FASTA file of CDS of
genes to add to reference
genome</p></td>
</tr>
<tr class="row-even"><td rowspan="8"><p>[seq]</p>
<p>target
definition</p>
</td>
<td><p>--gene</p></td>
<td><p>gene<strong>s</strong></p></td>
<td><p>gene IDs</p></td>
</tr>
<tr class="row-odd"><td><p>--cluster</p></td>
<td></td>
<td><p>cluster aliases</p></td>
</tr>
<tr class="row-even"><td><p>--indv</p></td>
<td></td>
<td><p>individuals to discover
targets in</p></td>
</tr>
<tr class="row-odd"><td><p>--target</p></td>
<td><p>target</p></td>
<td><p>FASTA file of sequences
to find gRNA in</p></td>
</tr>
<tr class="row-even"><td><p>--query</p></td>
<td><p>query*</p></td>
<td><p>FASTA file(s) to discover
targets in</p></td>
</tr>
<tr class="row-odd"><td><p>--domain &lt;alias&gt;</p></td>
<td></td>
<td><p>aliases of domains to
find gRNA in</p></td>
</tr>
<tr class="row-even"><td><p>--domain &lt;Pssm-Id&gt;</p></td>
<td><p>pssm_ids</p></td>
<td><p>Pssm-Id(s) of domains to
find gRNA in</p></td>
</tr>
<tr class="row-odd"><td></td>
<td><p>domain_name</p></td>
<td><p>human-readable domain
name used in sequence and
file names in place of
Pssm-Ids</p></td>
</tr>
<tr class="row-even"><td rowspan="7"><p>[seq]</p>
<p>inferring
homologues</p>
<p>from BLASTN
hits</p>
</td>
<td><p>--minid</p></td>
<td><p>minid</p></td>
<td><p>minimum hit % identity</p></td>
</tr>
<tr class="row-odd"><td><p>--minlen</p></td>
<td><p>minlen</p></td>
<td><p>minimum merged hits
length</p></td>
</tr>
<tr class="row-even"><td><p>--mincdslen</p></td>
<td><p>mincdslen</p></td>
<td><p>minimum CDS length of
merged hits</p></td>
</tr>
<tr class="row-odd"><td><p>--check-recip</p></td>
<td><p>check_recip</p></td>
<td><p>execute reciprocal check</p></td>
</tr>
<tr class="row-even"><td><p>--relax-recip</p></td>
<td><p>relax_recip</p></td>
<td><p>execute relaxed
reciprocal check</p></td>
</tr>
<tr class="row-odd"><td><p>--merge-within</p></td>
<td><p>merge_within</p></td>
<td><p>maximum distance between
hits for merging</p></td>
</tr>
<tr class="row-even"><td><p>--check-id-before-merge</p></td>
<td><p>check_id_before_merge</p></td>
<td><p>filter hits by % identity
before merging</p></td>
</tr>
<tr class="row-odd"><td rowspan="2"><p>[seq]</p>
<p>RPS-BLAST
options</p>
</td>
<td><p>--db</p></td>
<td><p>db</p></td>
<td><p>path to local RPS-BLAST
database</p></td>
</tr>
<tr class="row-even"><td><p>--remote-rps</p></td>
<td><p>remote_rps</p></td>
<td><p>use remote RPS-BLAST
database (currently
non-functional)</p></td>
</tr>
<tr class="row-odd"><td rowspan="2"><p>[grna]</p></td>
<td><p>--<a class="reference internal" href="#pam"><span class="std std-ref">pam</span></a></p></td>
<td><p><a class="reference internal" href="#pam"><span class="std std-ref">pam</span></a></p></td>
<td><p>PAM pattern</p></td>
</tr>
<tr class="row-even"><td><p>--length</p></td>
<td><p>length</p></td>
<td><p>gRNA length</p></td>
</tr>
<tr class="row-odd"><td rowspan="2"><p>[filter]</p>
<p>GC</p>
</td>
<td><p>--gc-min</p></td>
<td><p>gc_min</p></td>
<td><p>minimum GC content</p></td>
</tr>
<tr class="row-even"><td><p>--gc-max</p></td>
<td><p>gc_max</p></td>
<td><p>maximum GC content</p></td>
</tr>
<tr class="row-odd"><td rowspan="4"><p>[filter]</p>
<p>feature</p>
</td>
<td><p>--feature</p></td>
<td><p>feature</p></td>
<td><p>GFF3 feature type</p></td>
</tr>
<tr class="row-even"><td><p>--max-insertion</p></td>
<td><p>max_insertion</p></td>
<td><p>maximum allowable
insertion in feature</p></td>
</tr>
<tr class="row-odd"><td><p>--min-within-n</p></td>
<td><p>min_within_n</p></td>
<td><p>minimum number of
reference genes which
features overlap with
gRNA range in alignment</p></td>
</tr>
<tr class="row-even"><td><p>--min-within-fraction</p></td>
<td><p>min_within_fraction</p></td>
<td><p>minimum fraction of
reference genes which
features overlap with
gRNA range in alignment</p></td>
</tr>
<tr class="row-odd"><td rowspan="15"><p>[filter]</p>
<p>background</p>
</td>
<td><p>--background</p></td>
<td><p>background*</p></td>
<td><p>FASTA files in which to
search for potential
off-targets</p></td>
</tr>
<tr class="row-even"><td><p>--screen-reference</p></td>
<td><p>screen_reference</p></td>
<td><p>include reference genomes
in search for potential
off-targets</p></td>
</tr>
<tr class="row-odd"><td></td>
<td><p>mask</p></td>
<td><p>FASTA files of additional
sequences to mask</p></td>
</tr>
<tr class="row-even"><td><p>--unmask-ref</p></td>
<td></td>
<td><p>unmask reference genes</p></td>
</tr>
<tr class="row-odd"><td><p>--mask-gene</p></td>
<td></td>
<td><p>additional genes to mask</p></td>
</tr>
<tr class="row-even"><td><p>--unmask-gene</p></td>
<td></td>
<td><p>genes to unmask</p></td>
</tr>
<tr class="row-odd"><td><p>--mask-cluster</p></td>
<td></td>
<td><p>additional clusters to
mask</p></td>
</tr>
<tr class="row-even"><td><p>--unmask-cluster</p></td>
<td></td>
<td><p>clusters to unmask</p></td>
</tr>
<tr class="row-odd"><td><p>--ot-pamless</p></td>
<td><p>ot_pamless</p></td>
<td><p>ignore absense of PAM for
potential off-targets</p></td>
</tr>
<tr class="row-even"><td><p>--ot-mismatch</p></td>
<td><p>ot_mismatch</p></td>
<td><p>minimum acceptable
mismatches for
off-targets</p></td>
</tr>
<tr class="row-odd"><td><p>--ot-gap</p></td>
<td><p>ot_gap</p></td>
<td><p>minimum acceptable gaps
for off-targets</p></td>
</tr>
<tr class="row-even"><td><p>--<a class="reference internal" href="#off-target-pattern"><span class="std std-ref">ot-pattern</span></a></p></td>
<td><p><a class="reference internal" href="#off-target-pattern"><span class="std std-ref">ot_pattern</span></a></p></td>
<td><p>pattern to define
combination, number, and
location of gap(s) and/or
mismatch(es) for
unacceptable off-target
hits (i.e. gRNA with
off-target hits that
match the defined pattern
will be excluded)</p></td>
</tr>
<tr class="row-odd"><td><p>--ot-unaligned-as-mismatch</p></td>
<td><p>ot_unaligned_as_mismatch</p></td>
<td><p>treat unaligned positions
as mismatches (used with
--ot-pattern/ot_pattern)</p></td>
</tr>
<tr class="row-even"><td><p>--ot-unaligned-as-gap</p></td>
<td><p>ot_unaligned_as_gap</p></td>
<td><p>treat unaligned positions
as gaps (used with
--ot-pattern/ot_pattern)</p></td>
</tr>
<tr class="row-odd"><td><p>--skip-bg-check</p></td>
<td></td>
<td><p>skip off-target check</p></td>
</tr>
<tr class="row-even"><td><p>[filter]
exclude</p></td>
<td><p>--exclude</p></td>
<td><p>exclude</p></td>
<td><p>FASTA file of gRNA
sequences to exclude</p></td>
</tr>
<tr class="row-odd"><td rowspan="6"><p>[minimumset]</p></td>
<td><p>--accept-invalid</p></td>
<td><p>accept_invalid</p></td>
<td><p>score ‘NA’ as ‘pass’</p></td>
</tr>
<tr class="row-even"><td><p>--accept-feature-unknown</p></td>
<td><p>accept_feature_unknown</p></td>
<td><p>score ‘NA’ as ‘pass’ for
feature check</p></td>
</tr>
<tr class="row-odd"><td></td>
<td><p>accept_invalid_field</p></td>
<td><p>score ‘NA’ as ‘pass’ if
all entries for a check
are ‘NA’</p></td>
</tr>
<tr class="row-even"><td><p>--sets</p></td>
<td><p>sets</p></td>
<td><p>number of gRNA sets</p></td>
</tr>
<tr class="row-odd"><td><p>--auto</p></td>
<td><p>auto</p></td>
<td><p>generate sets without
require manual user
confirmation for each set</p></td>
</tr>
<tr class="row-even"><td><p>--prioritise-nr</p></td>
<td><p>prioritise_nr</p></td>
<td><p>prioritise non-redundancy
(nr) over proximity to 5’</p>
<blockquote>
<div><p>when selecting next gRNA</p>
</div></blockquote>
<p>in set</p>
</td>
</tr>
</tbody>
</table>
</section>
<section id="parameter-types">
<h2><a class="toc-backref" href="#id4"><span class="section-number">7.2. </span>Parameter types</a><a class="headerlink" href="#parameter-types" title="Permalink to this headline"></a></h2>
<section id="flags-and-arguments">
<h3><a class="toc-backref" href="#id5"><span class="section-number">7.2.1. </span>Flags and arguments</a><a class="headerlink" href="#flags-and-arguments" title="Permalink to this headline"></a></h3>
<section id="flag">
<h4><a class="toc-backref" href="#id6"><span class="section-number">7.2.1.1. </span>Flag</a><a class="headerlink" href="#flag" title="Permalink to this headline"></a></h4>
<p>Flags are parameters that do not take values.</p>
<p><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--auto</span></code>, <code class="docutils literal notranslate"><span class="pre">--accept-invalid</span></code>, <code class="docutils literal notranslate"><span class="pre">--accept-feature-unknown</span></code>, <code class="docutils literal notranslate"><span class="pre">--prioritise-nr</span></code>/<code class="docutils literal notranslate"><span class="pre">--prioritise-pos</span></code>, <code class="docutils literal notranslate"><span class="pre">--ot-unaligned-as-gap</span></code>/<code class="docutils literal notranslate"><span class="pre">--ot-uag</span></code>, <code class="docutils literal notranslate"><span class="pre">--ot-unaligned-as-mismatch</span></code>/<code class="docutils literal notranslate"><span class="pre">--ot-uam</span></code></p>
<p>For example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;other arguments&gt; --auto
</pre></div>
</div>
<p>Simply using <code class="docutils literal notranslate"><span class="pre">--auto</span></code> tells MINORg to automate set generation.</p>
<p><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">auto</span></code>, <code class="docutils literal notranslate"><span class="pre">accept_invalid</span></code>, <code class="docutils literal notranslate"><span class="pre">accept_feature_unknown</span></code>, <code class="docutils literal notranslate"><span class="pre">accept_invalid_field</span></code>, <code class="docutils literal notranslate"><span class="pre">prioritise_nr</span></code>/<code class="docutils literal notranslate"><span class="pre">prioritise_pos</span></code>, <code class="docutils literal notranslate"><span class="pre">unaligned_as_gap</span></code>, <code class="docutils literal notranslate"><span class="pre">unaligned_as_mismatch</span></code></p>
<p>In Python, flags are raised by setting the value of their attributes to <code class="docutils literal notranslate"><span class="pre">True</span></code> or <code class="docutils literal notranslate"><span class="pre">False</span></code>. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">auto</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1">## raise flag for parameter &#39;auto&#39;</span>
</pre></div>
</div>
</section>
<section id="argument">
<h4><a class="toc-backref" href="#id7"><span class="section-number">7.2.1.2. </span>Argument</a><a class="headerlink" href="#argument" title="Permalink to this headline"></a></h4>
<p>These parameters take values.</p>
<p><strong>CLI</strong>: all parameters that are not flags</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;other arguments&gt; --prefix my_minorg
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--prefix</span> <span class="pre">my_minorg</span></code> tells MINORg to use ‘my_minorg’ as a prefix for output files and directories.</p>
<p><strong>Python</strong>: all parameters that are not flags</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="s1">&#39;my_minorg&#39;</span> <span class="c1">## tells MINORg to use &#39;my_minorg&#39; as prefix for output files and directories</span>
</pre></div>
</div>
</section>
</section>
<section id="paths">
<h3><a class="toc-backref" href="#id8"><span class="section-number">7.2.2. </span>Paths</a><a class="headerlink" href="#paths" title="Permalink to this headline"></a></h3>
<div class="line-block">
<div class="line"><strong>CLI</strong>: As all paths will be resolved to absolute paths, relative paths are acceptable. Nevertheless, do be careful with relative paths and NEVER use them in the config file or in lookup files.</div>
<div class="line"><strong>Python</strong>: Paths are NOT resolved (except directory and config file). Absolute paths are STRONGLY RECOMMENDED. Be careful with relative paths.</div>
</div>
<section id="executables">
<h4><a class="toc-backref" href="#id9"><span class="section-number">7.2.2.1. </span>Executables</a><a class="headerlink" href="#executables" title="Permalink to this headline"></a></h4>
<p>Default values for executables may be specified in the config file (see <a class="reference internal" href="configuration.html#configuration"><span class="std std-ref">Configuration</span></a> for more on the config file).</p>
<section id="blastn-rpsblast-rpsblast-mafft">
<h5><span class="section-number">7.2.2.1.1. </span>blastn, rpsblast/rpsblast+, MAFFT<a class="headerlink" href="#blastn-rpsblast-rpsblast-mafft" title="Permalink to this headline"></a></h5>
<p>If an executable is in the command-search path, specifying these parameters is optional, although you may, if you desire, specify the command itself (e.g. ‘blastn’ instead of ‘/usr/bin/blastn’). If not, the <strong>path to the executable</strong> is required.</p>
<p>To determine if blastn and rpsblast (or rpsblast+ depending on your BLAST+ version) in the command-search path, execute at the command line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">blastn</span> <span class="o">-</span><span class="n">version</span>
</pre></div>
</div>
<p>If it prints something like</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">blastn</span><span class="p">:</span> <span class="mf">2.6.0</span><span class="o">+</span>
 <span class="n">Package</span><span class="p">:</span> <span class="n">blast</span> <span class="mf">2.6.0</span><span class="p">,</span> <span class="n">build</span> <span class="n">Jan</span> <span class="mi">15</span> <span class="mi">2017</span> <span class="mi">17</span><span class="p">:</span><span class="mi">12</span><span class="p">:</span><span class="mi">27</span>
</pre></div>
</div>
<p>then ‘blastn’ IS in your command-search path. Repeat this with ‘rpsblast’ and/or ‘rpsblast+’.</p>
<p>To determine if the mafft is in your command-search path, execute at the command line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mafft</span> <span class="o">--</span><span class="n">version</span>
</pre></div>
</div>
<p>If it prints something like</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">v7</span><span class="mf">.427</span> <span class="p">(</span><span class="mi">2019</span><span class="o">/</span><span class="n">Mar</span><span class="o">/</span><span class="mi">24</span><span class="p">)</span>
</pre></div>
</div>
<p>then it IS in your command-search path.</p>
<p><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--blastn</span></code>, <code class="docutils literal notranslate"><span class="pre">--rpsblast</span></code>, <code class="docutils literal notranslate"><span class="pre">--maff</span></code></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;other arguments&gt; --blastn /usr/bin/blastn
</pre></div>
</div>
<p><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">blastn</span></code>, <code class="docutils literal notranslate"><span class="pre">rpsblast</span></code>, <code class="docutils literal notranslate"><span class="pre">mafft</span></code></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">blastn</span> <span class="o">=</span> <span class="s1">&#39;/usr/bin/blastn&#39;</span> <span class="c1">## tells MINORg where the blastn executable is</span>
</pre></div>
</div>
</section>
<section id="bedtools">
<h5><span class="section-number">7.2.2.1.2. </span>BEDTools<a class="headerlink" href="#bedtools" title="Permalink to this headline"></a></h5>
<p>If bedtools is in the command-search path, you should NOT use this parameter. If not, the path to the <strong>directory containing the BEDTools executables</strong> is required.</p>
<p>To determine if the BEDTools executables are in your command-search path, execute at the command line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bedtools</span> <span class="o">--</span><span class="n">version</span>
</pre></div>
</div>
<p>If it prints something like</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bedtools</span> <span class="n">v2</span><span class="mf">.26.0</span>
</pre></div>
</div>
<p>then ‘bedtools’ is in your command-search path.</p>
<p><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">bedtools</span></code></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;other arguments&gt; --bedtools /path/to/bedtools2/bin/
</pre></div>
</div>
<p><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">bedtools</span></code></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">bedtools</span> <span class="o">=</span> <span class="s1">&#39;/path/to/bedtools2/bin/&#39;</span> <span class="c1">## tells MINORg where the BEDTools executables are</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="alias-lookup">
<h3><a class="toc-backref" href="#id10"><span class="section-number">7.2.3. </span>Alias lookup</a><a class="headerlink" href="#alias-lookup" title="Permalink to this headline"></a></h3>
<p>Note that aliases are <strong>case-sensitive</strong>.</p>
<section id="level-lookup">
<h4><a class="toc-backref" href="#id11"><span class="section-number">7.2.3.1. </span>1-level lookup</a><a class="headerlink" href="#level-lookup" title="Permalink to this headline"></a></h4>
<p>See also: <a class="reference internal" href="configuration.html#level-lookup"><span class="std std-ref">1-level lookup</span></a></p>
<p>1-level lookup parameters have preset values mapped to aliases defined in a configuration file. Users may use either the alias(es) or provide raw values.</p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--assembly</span></code>, <code class="docutils literal notranslate"><span class="pre">--annotation</span></code>, <code class="docutils literal notranslate"><span class="pre">--db</span></code>, <code class="docutils literal notranslate"><span class="pre">attr-mod</span></code>, <code class="docutils literal notranslate"><span class="pre">--domain</span></code></div>
<div class="line"><strong>Python</strong>: Does not support aliases. Raw values only.</div>
</div>
</section>
<section id="id1">
<h4><a class="toc-backref" href="#id12"><span class="section-number">7.2.3.2. </span>2-level lookup</a><a class="headerlink" href="#id1" title="Permalink to this headline"></a></h4>
<p>See also: <a class="reference internal" href="configuration.html#id1"><span class="std std-ref">2-level lookup</span></a></p>
<p>2-level lookup parameters use a combination of 2 parameters. The first parameter (suffixed with <cite>set</cite>) specifies a file containing alias mapping information for the second parameter (not suffixed). Aliases for the first parameter are defined in a configuration file, and functions effectively the same way a 1-level lookup parameter does. The second parameter reads alias mapping information from the file specified by the first parameter. Unlike the first parameters, users may only use alias(es)–raw values are not allowed. To specify raw values, different parameters must be used (see <a class="reference internal" href="configuration.html#alternative-parameters"><span class="std std-ref">Alternative parameters</span></a> for which).</p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code>-<code class="docutils literal notranslate"><span class="pre">reference</span></code>, <code class="docutils literal notranslate"><span class="pre">--cluster-set</span></code>-<code class="docutils literal notranslate"><span class="pre">--cluster</span></code>, <code class="docutils literal notranslate"><span class="pre">--genome-set</span></code>-<code class="docutils literal notranslate"><span class="pre">--indv</span></code></div>
<div class="line"><strong>Python</strong>: Does not support aliases. Raw values only.</div>
</div>
</section>
<section id="predefined-lookup">
<h4><a class="toc-backref" href="#id13"><span class="section-number">7.2.3.3. </span>Predefined lookup</a><a class="headerlink" href="#predefined-lookup" title="Permalink to this headline"></a></h4>
<p>Predefined lookup parameters are built into the programme. Users may use either the alias(es) or raw values.</p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--pam</span></code></div>
<div class="line"><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">pam</span></code></div>
</div>
</section>
<section id="raw-values">
<h4><a class="toc-backref" href="#id14"><span class="section-number">7.2.3.4. </span>Raw values</a><a class="headerlink" href="#raw-values" title="Permalink to this headline"></a></h4>
<p>All other parameters are raw values only.</p>
</section>
</section>
<section id="multiple-values">
<h3><a class="toc-backref" href="#id15"><span class="section-number">7.2.4. </span>Multiple values</a><a class="headerlink" href="#multiple-values" title="Permalink to this headline"></a></h3>
<section id="comma-separated-cli">
<h4><a class="toc-backref" href="#id16"><span class="section-number">7.2.4.1. </span>Comma-separated (CLI)</a><a class="headerlink" href="#comma-separated-cli" title="Permalink to this headline"></a></h4>
<p><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--reference</span></code>, <code class="docutils literal notranslate"><span class="pre">--cluster</span></code>, <code class="docutils literal notranslate"><span class="pre">--gene</span></code>, <code class="docutils literal notranslate"><span class="pre">--indv</span></code></p>
<p>Comma-separated multiple value arguments accept multiple values for a single argument so long as the values are comma-separated. For example, multiple genes can be specified using <code class="docutils literal notranslate"><span class="pre">--gene</span> <span class="pre">'geneA,geneB,geneC'</span></code>.</p>
</section>
<section id="multi-argument-cli">
<h4><a class="toc-backref" href="#id17"><span class="section-number">7.2.4.2. </span>Multi-argument (CLI)</a><a class="headerlink" href="#multi-argument-cli" title="Permalink to this headline"></a></h4>
<p><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--reference</span></code>, <code class="docutils literal notranslate"><span class="pre">--cluster</span></code>, <code class="docutils literal notranslate"><span class="pre">--gene</span></code>, <code class="docutils literal notranslate"><span class="pre">--indv</span></code>, <code class="docutils literal notranslate"><span class="pre">--query</span></code>, <code class="docutils literal notranslate"><span class="pre">--feature</span></code>, <code class="docutils literal notranslate"><span class="pre">--ext-gene</span></code>, <code class="docutils literal notranslate"><span class="pre">--ext-cds</span></code>, <code class="docutils literal notranslate"><span class="pre">--mask-gene</span></code>, <code class="docutils literal notranslate"><span class="pre">--unmask-gene</span></code>, <code class="docutils literal notranslate"><span class="pre">--mask-cluster</span></code>, <code class="docutils literal notranslate"><span class="pre">--unmask-cluster</span></code>
.. , <code class="docutils literal notranslate"><span class="pre">--ot-indv</span></code> (not implemented)</p>
<p>Multi-argument parameters accept multiple values by re-using a parameter. For example, multiple genes can be specified using <code class="docutils literal notranslate"><span class="pre">--gene</span> <span class="pre">geneA</span> <span class="pre">--gene</span> <span class="pre">geneB</span> <span class="pre">--gene</span> <span class="pre">geneC</span></code>.</p>
<p>(Note that some parameters can be both comma-separated AND multi-argument, and that these features can be combined. For example, <code class="docutils literal notranslate"><span class="pre">--gene</span> <span class="pre">geneA</span> <span class="pre">--gene</span> <span class="pre">geneB,geneC</span></code> is also valid.)</p>
</section>
<section id="multi-value-list-python">
<h4><a class="toc-backref" href="#id18"><span class="section-number">7.2.4.3. </span>Multi-value list (Python)</a><a class="headerlink" href="#multi-value-list-python" title="Permalink to this headline"></a></h4>
<p><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">genes</span></code></p>
<p>Multiple values for a single parameter may be provided to MINORg in a list. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;geneA&#39;</span><span class="p">]</span> <span class="c1">## specify a single value</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;geneA&#39;</span><span class="p">,</span> <span class="s1">&#39;geneB&#39;</span><span class="p">,</span> <span class="s1">&#39;geneC&#39;</span><span class="p">]</span> <span class="c1">## specify multiple values</span>
</pre></div>
</div>
</section>
<section id="multi-value-dictionary-python">
<h4><a class="toc-backref" href="#id19"><span class="section-number">7.2.4.4. </span>Multi-value dictionary (Python)</a><a class="headerlink" href="#multi-value-dictionary-python" title="Permalink to this headline"></a></h4>
<p><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">query</span></code>, <code class="docutils literal notranslate"><span class="pre">background</span></code></p>
<p>Multiple values for a single parameter may be provided to MINORg in a dictionary. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;queryA&#39;</span><span class="p">:</span> <span class="s1">&#39;/path/to/query_file.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;queryB&#39;</span><span class="p">:</span> <span class="s1">&#39;/path/to/another/query_file.fasta&#39;</span><span class="p">}</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="parameter-descriptions">
<h2><a class="toc-backref" href="#id20"><span class="section-number">7.3. </span>Parameter descriptions</a><a class="headerlink" href="#parameter-descriptions" title="Permalink to this headline"></a></h2>
<section id="reference">
<h3><a class="toc-backref" href="#id21"><span class="section-number">7.3.1. </span>Reference</a><a class="headerlink" href="#reference" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#argument"><span class="std std-ref">Argument</span></a>, <a class="reference internal" href="#id1"><span class="std std-ref">2-level lookup</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--reference</span></code> (used with <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code>)</div>
<div class="line"><strong>Python</strong>: set using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a>, get using <code class="docutils literal notranslate"><span class="pre">reference</span></code></div>
<div class="line"><strong>Config file</strong>:</div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">reference</span></code> (section <code class="docutils literal notranslate"><span class="pre">[data]</span></code>)</div>
<div class="line">set default set: <code class="docutils literal notranslate"><span class="pre">reference</span> <span class="pre">set</span></code> (section <code class="docutils literal notranslate"><span class="pre">[data]</span></code>)</div>
<div class="line">assign aliases to sets: <code class="docutils literal notranslate"><span class="pre">reference</span> <span class="pre">sets</span></code> (section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>)</div>
</div>
</div></blockquote>
<p>This paramter allows users to specify multiple reference genomes.</p>
<section id="reference-cli">
<h4><a class="toc-backref" href="#id22"><span class="section-number">7.3.1.1. </span>Reference (CLI)</a><a class="headerlink" href="#reference-cli" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="tutorial_cli.html#multiple-reference-genomes"><span class="std std-ref">Multiple reference genomes</span></a> for usage.</p>
<p>The primary difference between using <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">&lt;alias(es)&gt;</span> <span class="pre">--reference-set</span> <span class="pre">&lt;reference</span> <span class="pre">lookup</span> <span class="pre">file&gt;</span></code> and <code class="docutils literal notranslate"><span class="pre">--assembly</span> <span class="pre">&lt;FASTA&gt;</span> <span class="pre">--annotation</span> <span class="pre">&lt;GFF3&gt;</span></code> is that you can specify multiple genomes. This is achieved by supplying a reference lookup file (which maps a reference alias to a combination of &lt;FASTA&gt;-&lt;GFF3&gt;-&lt;genetic code&gt;-&lt;GFF3 attribute modification&gt;) using <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code> (see <a class="reference internal" href="configuration.html#reference"><span class="std std-ref">reference</span></a> for lookup file format) as well as the alias(es) of refence genome(s) to use using <code class="docutils literal notranslate"><span class="pre">--reference</span></code>.</p>
</section>
<section id="reference-python">
<h4><a class="toc-backref" href="#id23"><span class="section-number">7.3.1.2. </span>Reference (Python)</a><a class="headerlink" href="#reference-python" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="tutorial_py.html#multiple-reference-genomes"><span class="std std-ref">Multiple reference genomes</span></a> for an example of how to use the dedicated method <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a> to specify reference genomes, and <a class="reference internal" href="tutorial_py.html#non-standard-reference"><span class="std std-ref">Non-standard reference</span></a> for how to specify genetic code and GFF3 attribute mdifications for non-standard genomes/annotations.</p>
</section>
</section>
<section id="attribute-modification">
<h3><a class="toc-backref" href="#id24"><span class="section-number">7.3.2. </span>Attribute modification</a><a class="headerlink" href="#attribute-modification" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#argument"><span class="std std-ref">Argument</span></a>, <a class="reference internal" href="#level-lookup"><span class="std std-ref">1-level lookup</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code></div>
<div class="line"><strong>Python</strong>: NA (see argument <code class="docutils literal notranslate"><span class="pre">attr_mod</span></code> of <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a> instead)</div>
<div class="line"><strong>Config file</strong>:</div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">gff</span> <span class="pre">attribute</span> <span class="pre">modification</span></code> (section <code class="docutils literal notranslate"><span class="pre">[data]</span></code>)</div>
<div class="line">assign aliases: <code class="docutils literal notranslate"><span class="pre">gff</span> <span class="pre">attribute</span> <span class="pre">modification</span> <span class="pre">presets</span></code> (section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>)</div>
</div>
</div></blockquote>
<p>This parameter tells MINORg how to map non-standard GFF3 field names to standard GFF3 field names. This feature was originally developed when I tried to retrieve sequences using the IRGSP-1.0 annotation for rice (<em>Oryza sativa</em> subsp. Nipponbare) and discovered that it uses ‘Locus_id’ instead of ‘Parent’ for mRNA annotations.</p>
<p>See <a class="reference external" href="http://gmod.org/wiki/GFF3">http://gmod.org/wiki/GFF3</a> for standard attribute field names (see section titled ‘Column 9: “attributes”’).</p>
<section id="attribute-modification-format-cli">
<h4><a class="toc-backref" href="#id25"><span class="section-number">7.3.2.1. </span>Attribute modification format (CLI)</a><a class="headerlink" href="#attribute-modification-format-cli" title="Permalink to this headline"></a></h4>
<p>The input given to <code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code> should follow this format (with quotes):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>‘&lt;feature type&gt;:&lt;standard&gt;=&lt;nonstandard&gt;,&lt;standard&gt;=&lt;nonstandard&gt;;&lt;feature type&gt;:&lt;standard&gt;=&lt;nonstandard&gt;’
</pre></div>
</div>
<p>Examples:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">--attr-mod</span> <span class="pre">'mRNA:Parent=Locus_id,ID=transcript_id;CDS:Parent=transcript_id'</span></code></dt><dd><p>‘Locus_id’ and ‘transcript_id’ are non-standard field names for
fields ‘Parent’ and ‘ID’ respectively for the feature type ‘mRNA’,
and ‘transcript_id’ is the non-standard name for the field ‘Parent’ for the feature type ‘CDS’.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">--attr-mod</span> <span class="pre">'all:ID=id'</span></code></dt><dd><p>‘id’ is the non-standard field name for the field ‘ID’ for all feature types.</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="attribute-modification-format-reference-lookup-file">
<h4><a class="toc-backref" href="#id26"><span class="section-number">7.3.2.2. </span>Attribute modification format (reference lookup file)</a><a class="headerlink" href="#attribute-modification-format-reference-lookup-file" title="Permalink to this headline"></a></h4>
<p>See <a class="reference internal" href="#attribute-modification-format-cli"><span class="std std-ref">Attribute modification format (CLI)</span></a>, except quotes are not required.</p>
</section>
<section id="attribute-modification-format-python">
<h4><a class="toc-backref" href="#id27"><span class="section-number">7.3.2.3. </span>Attribute modification format (Python)</a><a class="headerlink" href="#attribute-modification-format-python" title="Permalink to this headline"></a></h4>
<p>The input given to the <code class="docutils literal notranslate"><span class="pre">attr_mod</span></code> keyword argument of the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a> method should be a dictionary following the following format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;&lt;feature type&gt;&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;&lt;standard&gt;&#39;</span><span class="p">:</span> <span class="s1">&#39;&lt;nonstandard&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;&lt;standard&gt;&#39;</span><span class="p">:</span> <span class="s1">&#39;&lt;nonstandard&gt;&#39;</span><span class="p">},</span>
 <span class="s1">&#39;&lt;feature type&gt;&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;&lt;standard&gt;&#39;</span><span class="p">:</span> <span class="s1">&#39;&lt;nonstandard&gt;&#39;</span><span class="p">}}</span>
</pre></div>
</div>
<p>Examples:</p>
<blockquote>
<div><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">{'mRNA':</span> <span class="pre">{'Parent':</span> <span class="pre">'Locus_id',</span> <span class="pre">'ID':</span> <span class="pre">'transcript_id'},</span> <span class="pre">'CDS':</span> <span class="pre">{'Parent':</span> <span class="pre">'transcript_id'}}</span></code></dt><dd><p>‘Locus_id’ and ‘transcript_id’ are non-standard field names for
fields ‘Parent’ and ‘ID’ respectively for the feature type ‘mRNA’,
and ‘transcript_id’ is the non-standard name for the field ‘Parent’ for the feature type ‘CDS’.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">{'all':</span> <span class="pre">{'ID':</span> <span class="pre">'id'}}</span></code></dt><dd><p>‘id’ is the non-standard field name for the field ‘ID’ for all feature types.</p>
</dd>
</dl>
</div></blockquote>
</section>
</section>
<section id="extended-genome">
<h3><a class="toc-backref" href="#id28"><span class="section-number">7.3.3. </span>Extended genome</a><a class="headerlink" href="#extended-genome" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#argument"><span class="std std-ref">Argument</span></a>, <a class="reference internal" href="#raw-values"><span class="std std-ref">Raw values</span></a>, <a class="reference internal" href="#multi-argument-cli"><span class="std std-ref">Multi-argument (CLI)</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--extend-gene</span></code>, <code class="docutils literal notranslate"><span class="pre">--extend-cds</span></code></div>
<div class="line"><strong>Python</strong>: use <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.extend_reference" title="minorg.MINORg.MINORg.extend_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">extend_reference()</span></code></a></div>
</div>
<p>These parameters accept FASTA files and allow MINORg to infer coding regions (CDS) from genomic (<code class="docutils literal notranslate"><span class="pre">--extend-gene</span></code>; first positional argument of <code class="xref py py-meth docutils literal notranslate"><span class="pre">extend_reference()</span></code>) and CDS-only (<code class="docutils literal notranslate"><span class="pre">--extend-cds</span></code>; second positional argument of <code class="xref py py-meth docutils literal notranslate"><span class="pre">extend_reference()</span></code>) sequences. They should be used when you do not have a GFF3 annotation file for your desired genes, but DO have the above mentioned sequences. MINORg will align gene and CDS-only sequences using MAFFT to generate a GFF3 annotation file with inferred intron-exon boundaries. These genes will then be added to the reference genome <strong>and you can use their gene IDs as you would reference gene IDs</strong>. You may provide multiple files to each parameter–MINORg will process them all simultaneously.</p>
<p>For MINORg to map the CDS-only sequences to the correct gene sequences, CDS-only sequences should be named according to the the format: ‘&lt;gene ID&gt;.&lt;CDS ID&gt;’</p>
<p>For example, given the following CDS sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">geneA</span><span class="mf">.1</span>
<span class="n">ATGATGATGATGATGATGATGATGTAA</span>
<span class="o">&gt;</span><span class="n">geneA</span><span class="o">.</span><span class="n">two</span>
<span class="n">ATGATGATGATGATGATGATGTAA</span>
<span class="o">&gt;</span><span class="n">geneA</span><span class="o">.</span><span class="n">foo</span><span class="o">.</span><span class="n">bar</span>
<span class="n">ATGATGATGATGATGATGTAA</span>
<span class="o">&gt;</span><span class="n">geneB</span><span class="mf">.1</span>
<span class="n">ATGAAAAAAAAAAAAAAAAAATAA</span>
</pre></div>
</div>
<p>And the following gene sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">geneA</span>
<span class="n">ATGATGATGATGATGATGATGATGTAA</span>
<span class="o">&gt;</span><span class="n">geneA</span><span class="o">.</span><span class="n">foo</span>
<span class="n">ATGATGATGATGATGATGATGATGTAA</span>
<span class="o">&gt;</span><span class="n">geneB</span>
<span class="n">ATGAAAAAAAAAAAAAAAAAAAAAAAATAA</span>
</pre></div>
</div>
<p>CDS sequences <code class="docutils literal notranslate"><span class="pre">geneA.1</span></code> and <code class="docutils literal notranslate"><span class="pre">geneA.two</span></code> will be mapped to gene sequence <code class="docutils literal notranslate"><span class="pre">geneA</span></code>, <code class="docutils literal notranslate"><span class="pre">geneA.foo.bar</span></code> will be mapped to <code class="docutils literal notranslate"><span class="pre">geneA.foo</span></code>, and <code class="docutils literal notranslate"><span class="pre">geneB.1</span></code> will be mapped to <code class="docutils literal notranslate"><span class="pre">geneB</span></code>. Note that <code class="docutils literal notranslate"><span class="pre">geneA.1</span></code> and <code class="docutils literal notranslate"><span class="pre">geneA.two</span></code> will be treated as different isoforms of the gene <code class="docutils literal notranslate"><span class="pre">geneA</span></code>.</p>
</section>
<section id="pam">
<h3><a class="toc-backref" href="#id29"><span class="section-number">7.3.4. </span>PAM</a><a class="headerlink" href="#pam" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#argument"><span class="std std-ref">Argument</span></a>, <a class="reference internal" href="#predefined-lookup"><span class="std std-ref">Predefined lookup</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--pam</span></code></div>
<div class="line"><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">pam</span></code></div>
<div class="line"><strong>Config file</strong>:</div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">pam</span></code> (section <code class="docutils literal notranslate"><span class="pre">[grna]</span></code>)</div>
<div class="line">assign aliases: <code class="docutils literal notranslate"><span class="pre">pam</span> <span class="pre">alias</span></code> (section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>) (not yet implemented)</div>
</div>
</div></blockquote>
<p>By default, MINORg designs gRNA for SpCas9 systems (i.e. 3’ NGG PAM). You may specify other PAM patterns for non-SpCas9 systems using <code class="docutils literal notranslate"><span class="pre">--pam</span></code>. It is recommended that any PAM pattern that uses special characters be enclosed in quotes, as it may lead to unexpected behaviour otherwise at the terminal.</p>
<p>Under the hood, MINORg uses regex to match PAM sites. Therefore, it is in theory possible to utilise the full suite of Python regex syntax to customise your PAM pattern. Note that PAM is <strong>NOT case-sensitive</strong>. However, do take care to avoid using  <code class="docutils literal notranslate"><span class="pre">.</span></code> as a wildcard, as MINORg uses this character to determine where gRNA is relative to a PAM pattern.</p>
<section id="ambiguous-bases-and-repeats">
<h4><a class="toc-backref" href="#id30"><span class="section-number">7.3.4.1. </span>Ambiguous bases and repeats</a><a class="headerlink" href="#ambiguous-bases-and-repeats" title="Permalink to this headline"></a></h4>
<p>Unlike many gRNA designers, MINORg accepts ambiguous bases (see: <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/iupac.html">https://genome.ucsc.edu/goldenPath/help/iupac.html</a> for IUPAC codes) as well as variable number of repeats.</p>
<blockquote>
<div><p>Example: The pattern ‘R{1,2}T’ (where ‘R’ means ‘A’ or ‘G’, and {1,2} means either 1 to 2 repetitions
of the character right before it) will match ‘AT’, ‘GT’, ‘AAT’, ‘AGT’, ‘GAT’, and ‘GGT’.</p>
</div></blockquote>
</section>
<section id="spacers-and-3-or-5-pam">
<h4><a class="toc-backref" href="#id31"><span class="section-number">7.3.4.2. </span>Spacers and 3’ or 5’ PAM</a><a class="headerlink" href="#spacers-and-3-or-5-pam" title="Permalink to this headline"></a></h4>
<p>In the absence of ‘N’ in the PAM pattern, MINORg will assume 3’ PAM with 1 spacer base (such as in the 3’ ‘NGG’ of SpCas9). If a pattern includes an ‘N’ at either end, MINORg will assume that the gRNA is directly adjacent to the ‘N’ base of the pattern. To specify a 5’ PAM in the absence of ‘N’ in the PAM pattern, ‘.’ should be inserted where the gRNA is.</p>
<blockquote>
<div><p>Example 1: <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">.NGG</span></code> and <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">NGG</span></code> and <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">GG</span></code> are functionally identical.
The latter two will be expanded to the most explicit pattern: <code class="docutils literal notranslate"><span class="pre">.NGG</span></code>.</p>
<p>Example 2: If a CRISPR system uses ‘GG’ PAM with NO spacer ‘N’ base, the PAM pattern has to be
specified to MINORg as <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">.GG</span></code>. Otherwise, MINORg will insert a spacer ‘N’ base, giving rise
to the incorrect explicit pattern of <code class="docutils literal notranslate"><span class="pre">.NGG</span></code> instead.</p>
<p>Example 3: AacCas12b uses a 5’ PAM with the pattern ‘TTN’, which can be specified to MINORg as
<code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">TTN</span></code> or <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">TTN.</span></code>, where <code class="docutils literal notranslate"><span class="pre">.</span></code> indicates where the gRNA is.
<code class="docutils literal notranslate"><span class="pre">.</span></code> is optional as this PAM pattern (TTN) includes ‘N’ at the end.
Therefore, MINORg will infer a 5’ PAM.</p>
<p>Example 4: Cas12a uses a 5’ PAM with the pattern ‘TTTV’, which can be specified to MINORg as
<code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">TTTV.</span></code> or <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">'T{3}V.'</span></code>, where <code class="docutils literal notranslate"><span class="pre">.</span></code> indicates where the gRNA is.
As the PAM pattern does not include ‘N’, the gRNA position MUST be explicitly indicated using <code class="docutils literal notranslate"><span class="pre">.</span></code>.
If <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">TTTV</span></code> is (incorrectly) used, MINORg will default to a 3’ PAM AND add a spacer base,
expanding it to the undesired explicit pattern <code class="docutils literal notranslate"><span class="pre">.NTTTV</span></code> .</p>
</div></blockquote>
<p>For a PAM-less search, use: <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">.</span></code> or <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">'.'</span></code>.</p>
</section>
<section id="preset-pam-patterns">
<h4><a class="toc-backref" href="#id32"><span class="section-number">7.3.4.3. </span>Preset PAM patterns</a><a class="headerlink" href="#preset-pam-patterns" title="Permalink to this headline"></a></h4>
<p>MINORg comes with several preset PAM patterns for different CRISPR systems.</p>
<blockquote>
<div><p>For example: <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">SpCas9</span></code> and <code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">.NGG</span></code> are functionally identical.</p>
</div></blockquote>
<table class="docutils align-default">
<colgroup>
<col style="width: 19%" />
<col style="width: 24%" />
<col style="width: 57%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>alias(es)</strong></p></td>
<td><p><strong>PAM sequence</strong>
(explicit)</p></td>
<td><p><strong>Notes</strong></p></td>
</tr>
<tr class="row-even"><td><p>SpCas9 OR
spcas9</p></td>
<td><p>.NGG</p></td>
<td><p>default</p></td>
</tr>
<tr class="row-odd"><td><p>SaCas9T OR
sacas9t</p></td>
<td><p>.NGRRT</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>SaCas9N OR
sacas9n</p></td>
<td><p>.NGRRN</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>NmeCas9 OR
nmecas9</p></td>
<td><p>.NNNNGATT</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>CjCas9 OR
cjcas9</p></td>
<td><p>.NNNNRYAC</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>StCas9 OR
stcas9</p></td>
<td><p>.NNAGAAW</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Cas12a OR
cas12a</p></td>
<td><p>TTTV.</p></td>
<td><p>5’ PAM</p></td>
</tr>
<tr class="row-odd"><td><p>AacCas12b OR
aaccas12b</p></td>
<td><p>TTN.</p></td>
<td><p>5’ PAM</p></td>
</tr>
<tr class="row-even"><td><p>BhCas12b OR
bhcas12b</p></td>
<td><p>DTTN.</p></td>
<td><p>5’ PAM</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="off-target-pattern">
<h3><a class="toc-backref" href="#id33"><span class="section-number">7.3.5. </span>Off-target pattern</a><a class="headerlink" href="#off-target-pattern" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#argument"><span class="std std-ref">Argument</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span></code></div>
<div class="line"><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">ot_pattern</span></code></div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">off-target</span> <span class="pre">pattern</span></code> (section <code class="docutils literal notranslate"><span class="pre">[filter]</span></code>)</div>
</div>
</div></blockquote>
<p>For greater flexibility, MINORg provides a method for defining position-specific tolerances for gaps/mismatches/unaligned positions.</p>
<p>(By default, MINORg uses <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code>/<code class="docutils literal notranslate"><span class="pre">ot_mismatch</span></code> and <code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code>/<code class="docutils literal notranslate"><span class="pre">ot_gap</span></code> to determine whether an off-target hit disqualifies as gRNA. This default behaviour counts the total number of mismatches and/or gaps and/or unaligned positions in an off-target gRNA hit and discards or retains a gRNA based on the specified threshold values. See <a class="reference internal" href="algorithms.html#total-mismatch-gap-unaligned"><span class="std std-ref">Total mismatch/gap/unaligned</span></a> for this default algorithm. This behaviour will be overridden if <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span></code>/<code class="docutils literal notranslate"><span class="pre">ot_pattern</span></code> is specified.)</p>
<section id="basic-unit">
<h4><a class="toc-backref" href="#id34"><span class="section-number">7.3.5.1. </span>Basic unit</a><a class="headerlink" href="#basic-unit" title="Permalink to this headline"></a></h4>
<p>The basic unit of an off-target pattern comprises of 3 parts:</p>
<ul class="simple">
<li><p>Maximum intolerable count (integer)</p></li>
<li><p>Type of non-match (whether gap, deletion, insertion, and/or gap)</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">m</span></code>: mismatch</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">g</span></code>: gap (should not be used with <code class="docutils literal notranslate"><span class="pre">i</span></code> and/or <code class="docutils literal notranslate"><span class="pre">d</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">i</span></code>: insertion (base present in gRNA but not in the off-target sequence)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">d</span></code>: deletion (base not present in gRNA but present in the off-target sequence)</p></li>
</ul>
</li>
<li><p>Range</p>
<ul>
<li><p>All examples below will be based on a very short 8 bp gRNA of 5’-ATGCatgc-3’ (upper and lowercase for illustration purposes)</p></li>
<li><p>Position indices can be positive or negative, but not zero.</p>
<ul>
<li><p>This allows flexibility regardless of gRNA length and whether PAM is 5’ or 3’.</p></li>
<li><p>If index &gt; 0: positions are counted from the 5’ end (best for 5’ PAM)</p>
<ul>
<li><p>Index 1 = A</p></li>
<li><p>Index 5 = a</p></li>
<li><p>Index 7 = g</p></li>
</ul>
</li>
<li><p>If index &lt; 0: positions are counted from the 3’ end (best for 3’ PAM)</p>
<ul>
<li><p>Index -1 = c</p></li>
<li><p>Index -5 = C</p></li>
<li><p>Index -7 = T</p></li>
</ul>
</li>
</ul>
</li>
<li><p>If a single index is provided, the range is assumed to be:</p>
<ul>
<li><p>&lt;start&gt; to &lt;index&gt;: if index &gt; 0</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">4</span></code>: ATGC (positions 1 to 4)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">6</span></code>: ATGCat (positions 1 to 6)</p></li>
</ul>
</li>
<li><p>&lt;index&gt; to &lt;end&gt;: if index &lt; 0</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-4</span></code>: atgc (positions -4 to -1)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-6</span></code>: GCatgc (positions -6 to -1)</p></li>
</ul>
</li>
</ul>
</li>
<li><p>If a single index is provided AND followed by a ‘-’, the range is assumed to be:</p>
<ul>
<li><p>&lt;index&gt; to &lt;end&gt;: if index &gt; 0</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">4-</span></code>: Catgc (positions 4 to 8)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">6-</span></code>: tgc (positions 6 to 8)</p></li>
</ul>
</li>
<li><p>&lt;start&gt; to &lt;index&gt;: if index &lt; 0</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-4-</span></code>: ATGCa (positions -8 to -4)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-6-</span></code>: ATG (positions -8 to -6)</p></li>
</ul>
</li>
</ul>
</li>
<li><p>Otherwise, a range can be defined using 2 indices separated by ‘-’. Values must either both be positive or both be negative. For ranges defined by negative indices, the smaller absolute value should come first.</p>
<ul>
<li><p>Valid</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">2-5</span></code>: TGCa (positions 2 to 5)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-2--5</span></code>: Catg (positions -5 to -2)</p></li>
</ul>
</li>
<li><p>Invalid</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">2--5</span></code>: mixed signs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-2-5</span></code>: mixed signs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-5--2</span></code>: smaller absolute value should come first</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<section id="examples">
<h5><span class="section-number">7.3.5.1.1. </span>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">0mg5</span></code>: gRNA hit with any mismatches or gaps (&gt;0) from positions 1 to 5 will be NOT be considered problematic.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">1i-5--20</span></code>: gRNA hit with more than 1 (&gt;1) insertions from positions -5 to -20 will NOT be considered problematic.</p></li>
</ul>
</section>
</section>
<section id="operators">
<h4><a class="toc-backref" href="#id35"><span class="section-number">7.3.5.2. </span>Operators</a><a class="headerlink" href="#operators" title="Permalink to this headline"></a></h4>
<p>Multiple units can be combined using <code class="docutils literal notranslate"><span class="pre">,</span></code> (AND) and <code class="docutils literal notranslate"><span class="pre">|</span></code> (OR).</p>
<p>Neither operator is prioritised over the other. You may specify order using parentheses <code class="docutils literal notranslate"><span class="pre">(</span></code> and <code class="docutils literal notranslate"><span class="pre">)</span></code>. In the absence of parenthesis, operations are evaluated left to right.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">0mg5,1mg6-|0mg6,1m7-</span></code> will be evaluated as <code class="docutils literal notranslate"><span class="pre">(((0mg5,1mg6-)|0mg6),1m7-)</span></code></p>
<ul>
<li><p>To evaluate <code class="docutils literal notranslate"><span class="pre">0mg5,1mg6-|0mg6,1m7-</span></code> as ‘<code class="docutils literal notranslate"><span class="pre">0mg5,1mg6-</span></code> OR <code class="docutils literal notranslate"><span class="pre">0mg6,1m7-</span></code>‘, use <code class="docutils literal notranslate"><span class="pre">(0mg5,1mg6-)|(0mg6,1m7-)</span></code></p></li>
</ul>
</li>
</ul>
<p>NOTE: You can technically combine basic units with ranges that are negative and positive (e.g <code class="docutils literal notranslate"><span class="pre">0mg5,0mg-5</span></code> is valid), but I’m not sure why you’d do that.</p>
<section id="id2">
<h5><span class="section-number">7.3.5.2.1. </span>Examples<a class="headerlink" href="#id2" title="Permalink to this headline"></a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">(0mg5,1mg6-)|(0mg6,1m7-)</span></code>: gRNA hit with &lt;no gaps/mismatches from positions 1 to 5 and no more than 1 gaps/mismatches from positions 6 to the end&gt; OR &lt;no gaps/mismatches from positions 1 to 6 and no more than 1 mismatch from positions 7 to the end REGARDLESS OF HOW MANY GAPS&gt; will be considered <strong>problematic</strong>.</p></li>
</ul>
</section>
</section>
</section>
<section id="prioritise-non-redundancy">
<h3><a class="toc-backref" href="#id36"><span class="section-number">7.3.6. </span>Prioritise non-redundancy</a><a class="headerlink" href="#prioritise-non-redundancy" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#flag"><span class="std std-ref">Flag</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--prioritise-nr</span></code>/<code class="docutils literal notranslate"><span class="pre">--prioritize-nr</span></code></div>
<div class="line"><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">prioritise_nr</span></code>/<code class="docutils literal notranslate"><span class="pre">--prioritize-nr</span></code></div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">prioritise</span> <span class="pre">non-redundnacy</span></code> (section <code class="docutils literal notranslate"><span class="pre">[filter]</span></code>)</div>
</div>
</div></blockquote>
<p>By default, gRNA are selected for a set in the following order of priority:</p>
<ol class="arabic simple">
<li><p>Coverage
- Favour gRNA that cover a larger number of targets not covered by already selected gRNA</p></li>
<li><p>Proximity to 5’
- Favour gRNA that are positioned closer to the 5’ end of a target
- For reference genes, MINORg favours proxiity to the 5’ end of the <strong>sense strand</strong>
- If reference genes have been specified, an alignment would have been generated with targets and reference genes, and sense will be inferred from this alignment. With sense information, MINORg will favour proximity to the 5’ end of the <strong>sense strand</strong>.</p></li>
<li><p>Non-redundancy
- Favour gRNA which coverage has the fewest overlap with targets covered by already selected gRNA</p></li>
</ol>
<p>If this flag is raised, ‘Non-redundancy’ will be prioritised before ‘Proximity to 5’. This may be preferred if you wish to generate a large number of sets, as priortisation of non-redundancy makes it less likely that extremely high coverage gRNA will be added to a growing set, such that these gRNA can then be used to seed the next set.</p>
</section>
<section id="rps-blast-local-database">
<h3><a class="toc-backref" href="#id37"><span class="section-number">7.3.7. </span>RPS-BLAST local database</a><a class="headerlink" href="#rps-blast-local-database" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#argument"><span class="std std-ref">Argument</span></a>, <a class="reference internal" href="#level-lookup"><span class="std std-ref">1-level lookup</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--db</span></code></div>
<div class="line"><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">db</span></code></div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">rps</span> <span class="pre">database</span></code> (section <code class="docutils literal notranslate"><span class="pre">[data]</span></code>)</div>
<div class="line">assign aliases: <code class="docutils literal notranslate"><span class="pre">rps</span> <span class="pre">database</span> <span class="pre">alias</span></code> (section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>)</div>
</div>
</div></blockquote>
<p>The latest CDD database may be downloaded at <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cdd.targ.gz">ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cdd.targ.gz</a>. As the CDD database is regularly updated, the PSSM-Id for a domain shown at the CDD website is subject to change. Thus, I also recommend downloading <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cddid.tbl.gz">ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cddid.tbl.gz</a>, which contains information that maps PSSM-Ids to domain accession IDs as well as domain names of the database version at the point of downloading.</p>
<p>Note: As the local database itself consists of multiple files with different extensions, the path provided to this parameter is not to any single file. For example, given the following file structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span>
<span class="o">+--</span> <span class="n">root</span><span class="o">/</span>
    <span class="o">|--</span> <span class="n">other_files</span><span class="o">/</span>
    <span class="o">+--</span> <span class="n">rps_db</span><span class="o">/</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">aux</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">freq</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">loo</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">phr</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">pin</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">psd</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">psi</span>
        <span class="o">|--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">psq</span>
        <span class="o">+--</span> <span class="n">Cdd</span><span class="o">.</span><span class="n">rps</span>
</pre></div>
</div>
<p>where the database is contained in the directory <code class="docutils literal notranslate"><span class="pre">/root/rsp_db/</span></code>, the appropriate path to pass to this parameter is: <code class="docutils literal notranslate"><span class="pre">/root/rps_db/Cdd</span></code>, where the trailing ‘Cdd’ is the prefix of all of the database’s files</p>
</section>
<section id="rps-blast-remote-database">
<h3><a class="toc-backref" href="#id38"><span class="section-number">7.3.8. </span>RPS-BLAST remote database</a><a class="headerlink" href="#rps-blast-remote-database" title="Permalink to this headline"></a></h3>
<p><strong>Type</strong>: <a class="reference internal" href="#flag"><span class="std std-ref">Flag</span></a></p>
<div class="line-block">
<div class="line"><strong>CLI</strong>: <code class="docutils literal notranslate"><span class="pre">--remote-rps</span></code></div>
<div class="line"><strong>Python</strong>: <code class="docutils literal notranslate"><span class="pre">remote_rps</span></code></div>
</div>
<blockquote>
<div><div class="line-block">
<div class="line">set default: <code class="docutils literal notranslate"><span class="pre">remote</span> <span class="pre">rps</span></code> (section <code class="docutils literal notranslate"><span class="pre">[data]</span></code>)</div>
</div>
</div></blockquote>
<p>While it is in theory possible to use the remote CDD database &amp; servers instead of local ones, the <code class="docutils literal notranslate"><span class="pre">--remote</span></code> option for the ‘rpsblast’/’rpsblast+’ command from the BLAST+ package has never worked for me. In any case, if your version of local rpsblast is able to access the remote database, you can use <code class="docutils literal notranslate"><span class="pre">--remote-rps</span></code> instead of <code class="docutils literal notranslate"><span class="pre">--db</span> <span class="pre">/path/to/rpsblast/db</span></code>.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial_py.html" class="btn btn-neutral float-left" title="6. Tutorial (Python)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="output.html" class="btn btn-neutral float-right" title="8. Output" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>