<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>5. Tutorial (Python) &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="6. Parameters" href="parameters.html" />
    <link rel="prev" title="4. Tutorial (Command line)" href="tutorial_cli.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">2. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">3. Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_cli.html">4. Tutorial (Command line)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">5. Tutorial (Python)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setting-up-the-tutorial">5.1. Setting up the tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-started">5.2. Getting started</a></li>
<li class="toctree-l2"><a class="reference internal" href="#impt-note-on-executables">5.3. IMPT: Note on executables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#defining-target-sequences">5.4. Defining target sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#user-provided-targets">5.4.1. User-provided targets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reference-gene-s-as-targets">5.4.2. Reference gene(s) as targets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#non-reference-gene-s-as-targets">5.4.3. Non-reference gene(s) as targets</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#extending-the-reference">5.4.3.1. Extending the reference</a></li>
<li class="toctree-l4"><a class="reference internal" href="#inferring-homologues-in-unannotated-genomes">5.4.3.2. Inferring homologues in unannotated genomes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#domain-as-targets">5.4.4. Domain as targets</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#local-database">5.4.4.1. Local database</a></li>
<li class="toctree-l4"><a class="reference internal" href="#remote-database">5.4.4.2. Remote database</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#defining-grna">5.5. Defining gRNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filtering-grna">5.6. Filtering gRNA</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#filter-by-gc-content">5.6.1. Filter by GC content</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filter-by-off-target">5.6.2. Filter by off-target</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pam-less-off-target-check">5.6.2.1. PAM-less off-target check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#skip-off-target-check">5.6.2.2. Skip off-target check</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filter-by-feature">5.6.3. Filter by feature</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#generating-minimum-grna-set-s">5.7. Generating minimum gRNA set(s)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#number-of-sets">5.7.1. Number of sets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#prioritise-non-redundancy">5.7.2. Prioritise non-redundancy</a></li>
<li class="toctree-l3"><a class="reference internal" href="#excluding-grna">5.7.3. Excluding gRNA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#accepting-unknown-checks">5.7.4. Accepting unknown checks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#manually-approve-grna-sets">5.7.5. Manually approve gRNA sets</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#subcommands">5.8. Subcommands</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-seq">5.8.1. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-grna">5.8.2. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">grna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-filter">5.8.3. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#subcommand-filter-gc">5.8.3.1. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_gc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#subcommand-filter-background">5.8.3.2. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_background()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#subcommand-filter-feature">5.8.3.3. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_feature()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-minimumset">5.8.4. Subcommand <code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#regenerating-minimum-sets-after-calling-full">5.8.4.1. Regenerating minimum sets after calling <code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#generating-minimum-sets-from-output-of-another-minorg-run">5.8.4.2. Generating minimum sets from output of another MINORg run</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#chaining-subcommands">5.8.5. Chaining subcommands</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#defining-reference-genomes">5.9. Defining reference genomes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#single-reference-genome">5.9.1. Single reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiple-reference-genomes">5.9.2. Multiple reference genomes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#non-standard-reference">5.9.3. Non-standard reference</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#non-standard-genetic-code">5.9.3.1. Non-standard genetic code</a></li>
<li class="toctree-l4"><a class="reference internal" href="#non-standard-gff3-attribute-field-names">5.9.3.2. Non-standard GFF3 attribute field names</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#multithreading">5.10. Multithreading</a></li>
<li class="toctree-l2"><a class="reference internal" href="#differences-between-cli-and-python-versions">5.11. Differences between CLI and Python versions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#to-specify-cluster-genes">5.11.1. To specify cluster genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#to-specify-query-fasta-files">5.11.2. To specify query FASTA files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#to-specify-reference-genomes">5.11.3. To specify reference genomes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">6. Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">7. Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">8. Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">9. Examples</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">5. </span>Tutorial (Python)</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorial_py.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial-python">
<h1><span class="section-number">5. </span>Tutorial (Python)<a class="headerlink" href="#tutorial-python" title="Permalink to this headline"></a></h1>
<p>Please download the files in <a class="reference external" href="https://github.com/rlrq/MINORg/tree/master/examples">https://github.com/rlrq/MINORg/tree/master/examples</a>. In all the examples below, you should replace “/path/to” with the appropriate full path name, which is usually to the directory containing these example files.</p>
<section id="setting-up-the-tutorial">
<h2><span class="section-number">5.1. </span>Setting up the tutorial<a class="headerlink" href="#setting-up-the-tutorial" title="Permalink to this headline"></a></h2>
<p>To ensure that the examples in this tutorial work, please replace ‘/path/to’ in the files ‘arabidopsis_genomes.txt’, ‘athaliana_genomes.txt’, and ‘subset_genome_mapping.txt’ with the full path to the directory containing the example files.</p>
</section>
<section id="getting-started">
<h2><span class="section-number">5.2. </span>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline"></a></h2>
<p>To begin, import the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a> class.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
</pre></div>
</div>
<p>To create a MINORg object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output/directory&quot;</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;prefix&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Both <code class="docutils literal notranslate"><span class="pre">directory</span></code> and <code class="docutils literal notranslate"><span class="pre">prefix</span></code> are optional. If not provided, they will default to the current directory and ‘minorg’ respectively. If the directory does not currently exist, it will be created.</p>
<p>If you wish to use the default values specified in a config file, use this instead:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">config</span> <span class="o">=</span> <span class="s2">&quot;/path/to/config.ini&quot;</span><span class="p">,</span> <span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output/directory&quot;</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;prefix&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You may now set your parameters using the attributes of your <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a> object. For a table listing the equivalent CLI arguments and <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a> attributes, see <a class="reference internal" href="parameters.html#cli-vs-python"><span class="std std-ref">CLI vs Python</span></a>.</p>
</section>
<section id="impt-note-on-executables">
<h2><span class="section-number">5.3. </span>IMPT: Note on executables<a class="headerlink" href="#impt-note-on-executables" title="Permalink to this headline"></a></h2>
<p>See: <a class="reference internal" href="parameters.html#executables"><span class="std std-ref">Executables</span></a></p>
<p>You can specify executables as such:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">blastn</span> <span class="o">=</span> <span class="s1">&#39;/path/to/blastn/executable&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">rpsblast</span> <span class="o">=</span> <span class="s1">&#39;/path/to/rpsblast/executable&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">mafft</span> <span class="o">=</span> <span class="s1">&#39;/path/to/mafft/executable&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">bedtools</span> <span class="o">=</span> <span class="s1">&#39;/path/to/bedtools2/bin&#39;</span>
</pre></div>
</div>
<p>Note that BEDTools is unique in that if it is not in your command-search path, you should provide the path <strong>TO THE DIRECTORY CONTAINING ITS EXECUTABLES</strong> (i.e. there will not be a single ‘bedtools’ executable), and if it <strong>IS</strong> in your command-search path you <strong>SHOULD NOT</strong> be using the <code class="xref py py-attr docutils literal notranslate"><span class="pre">bedtools</span></code> attribute. See <a class="reference internal" href="parameters.html#executables"><span class="std std-ref">Executables</span></a> for more on executables.</p>
</section>
<section id="defining-target-sequences">
<h2><span class="section-number">5.4. </span>Defining target sequences<a class="headerlink" href="#defining-target-sequences" title="Permalink to this headline"></a></h2>
<section id="user-provided-targets">
<h3><span class="section-number">5.4.1. </span>User-provided targets<a class="headerlink" href="#user-provided-targets" title="Permalink to this headline"></a></h3>
<p>Let us begin with the simplest MINORg execution:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_200_target&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>The above combination of arguments tells MINORg to generate gRNA from targets in a user-provided FASTA file (<code class="docutils literal notranslate"><span class="pre">my_minorg.target</span> <span class="pre">=</span> <span class="pre">'pat/to/sample_CDS.fasta'</span></code>) and to output files into the directory <code class="docutils literal notranslate"><span class="pre">/path/to/output/directory/example_200_target</span></code>. By default, MINORg generates 20 bp gRNA using NGG PAM. The full MINORg programme is executed by calling the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a> method (<code class="docutils literal notranslate"><span class="pre">my_minorg.full()</span></code>). Don’t forget to call <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.resolve" title="minorg.MINORg.MINORg.resolve"><code class="xref py py-meth docutils literal notranslate"><span class="pre">resolve()</span></code></a> to remove any temporary files.</p>
</section>
<section id="reference-gene-s-as-targets">
<h3><span class="section-number">5.4.2. </span>Reference gene(s) as targets<a class="headerlink" href="#reference-gene-s-as-targets" title="Permalink to this headline"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_201_refgene&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45060&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45200&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45210&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45220&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45230&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45240&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G45250&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the above example, <code class="docutils literal notranslate"><span class="pre">my_minorg.add_reference(&quot;/path/to/subset_ref_TAIR10.fasta&quot;,</span> <span class="pre">&quot;/path/to/subset_ref_TAIR10.gff&quot;,</span> <span class="pre">alias</span> <span class="pre">=</span> <span class="pre">&quot;TAIR10&quot;)</span></code> is used to specify information about a reference genome:</p>
<ul class="simple">
<li><p>Positional argument 1: path to reference assembly (In this case <code class="docutils literal notranslate"><span class="pre">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span></code>)</p></li>
<li><p>Positional argument 2: path to reference annotation (In this case <code class="docutils literal notranslate"><span class="pre">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span></code>)</p></li>
<li><p>Optional keyword argument 1 (<code class="docutils literal notranslate"><span class="pre">alias</span></code>): genome alias (in this case <code class="docutils literal notranslate"><span class="pre">&quot;TAIR10&quot;</span></code>); a unique name for the reference genome, used when referring to it in sequence names and output files. Autogenerated by MINORg if not provided.</p></li>
<li><p>See <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a> and <a class="reference internal" href="#non-standard-reference"><span class="std std-ref">Non-standard reference</span></a> for how to specify genetic code and non-standard attribute field names</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">my_minorg.genes</span> <span class="pre">=</span> <span class="pre">[&quot;AT5G45050&quot;,</span> <span class="pre">&quot;AT5G45060&quot;,</span> <span class="pre">&quot;AT5G45200&quot;,</span> <span class="pre">&quot;AT5G45210&quot;,</span> <span class="pre">&quot;AT5G45220&quot;,</span> <span class="pre">&quot;AT5G45230&quot;,</span> <span class="pre">&quot;AT5G45240&quot;,</span> <span class="pre">&quot;AT5G45250&quot;]</span></code> tells MINORg the target gene(s), and <code class="docutils literal notranslate"><span class="pre">my_minorg.query_reference</span> <span class="pre">=</span> <span class="pre">True</span></code> tells MINORg to generate gRNA for reference gene(s).</p>
</section>
<section id="non-reference-gene-s-as-targets">
<h3><span class="section-number">5.4.3. </span>Non-reference gene(s) as targets<a class="headerlink" href="#non-reference-gene-s-as-targets" title="Permalink to this headline"></a></h3>
<section id="extending-the-reference">
<h4><span class="section-number">5.4.3.1. </span>Extending the reference<a class="headerlink" href="#extending-the-reference" title="Permalink to this headline"></a></h4>
<p>See also: <a class="reference internal" href="parameters.html#extended-genome"><span class="std std-ref">Extended genome</span></a></p>
<p>If you have both genomic and CDS-only sequences of your target genes but not a GFF3 annotation file, MINORg can infer coding regions (CDS) for your target genes using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.extend_reference" title="minorg.MINORg.MINORg.extend_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">extend_reference()</span></code></a>. See <a class="reference internal" href="parameters.html#extended-genome"><span class="std std-ref">Extended genome</span></a> for how to name your sequences to ensure proper mapping of CDS to genes.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_202_ext&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">extend_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/sample_gene.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT1G10920&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.extend_reference" title="minorg.MINORg.MINORg.extend_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">extend_reference()</span></code></a> effectively adds new genes to the reference genome, so they can be used just like any reference gene. Therefore, they can also be used in combination with <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_query" title="minorg.MINORg.MINORg.add_query"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_query()</span></code></a>.</p>
</section>
<section id="inferring-homologues-in-unannotated-genomes">
<h4><span class="section-number">5.4.3.2. </span>Inferring homologues in unannotated genomes<a class="headerlink" href="#inferring-homologues-in-unannotated-genomes" title="Permalink to this headline"></a></h4>
<p>See also: <a class="reference internal" href="algorithms.html#non-reference-homologue-inference"><span class="std std-ref">Non-reference homologue inference</span></a></p>
<p>If you would like MINORg to infer homologues genes in non-reference genomes, you can use <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_query" title="minorg.MINORg.MINORg.add_query"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_query()</span></code></a> to specify the FASTA files of those non-reference genomes.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_203_query&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">extend_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/sample_gene.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT1G10920&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the above example, <code class="docutils literal notranslate"><span class="pre">my_minorg.add_query(&quot;/path/to/subset_9654.fasta&quot;,</span> <span class="pre">alias</span> <span class="pre">=</span> <span class="pre">&quot;9654&quot;)</span></code> and <code class="docutils literal notranslate"><span class="pre">my_minorg.add_query(&quot;/path/to/subset_9655.fasta&quot;,</span> <span class="pre">alias</span> <span class="pre">=</span> <span class="pre">&quot;9655&quot;)</span></code> are used to specify information about query FASTA files.</p>
<ul>
<li><p>The alias keyword argument is optional. If not provided, MINORg will generate a unique alias.</p></li>
<li><p>Query FASTA files are stored as a dictionary with the format {&lt;alias&gt;:&lt;FASTA&gt;} at <code class="xref py py-attr docutils literal notranslate"><span class="pre">query</span></code>.</p></li>
<li><p>If you’d like to remove a query file that you’ve added, you can use:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">remove_query</span><span class="p">(</span><span class="s2">&quot;9654&quot;</span><span class="p">)</span>
</pre></div>
</div>
<ul>
<li><p>The <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.remove_query" title="minorg.MINORg.MINORg.remove_query"><code class="xref py py-meth docutils literal notranslate"><span class="pre">remove_query()</span></code></a> method takes a query alias. If you did not specify an alias when using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_query" title="minorg.MINORg.MINORg.add_query"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_query()</span></code></a> and do not know the alias of the file you wish to remove, you may view the query-FASTA mapping using the <code class="xref py py-attr docutils literal notranslate"><span class="pre">query</span></code> attribute.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query</span>
<span class="go">{&quot;9654&quot;: &quot;/path/to/subset_9654.fasta&quot;, &quot;9655&quot;: &quot;/path/to/subset_9655.fasta&quot;}</span>
</pre></div>
</div>
</li>
</ul>
</li>
</ul>
</section>
</section>
<section id="domain-as-targets">
<h3><span class="section-number">5.4.4. </span>Domain as targets<a class="headerlink" href="#domain-as-targets" title="Permalink to this headline"></a></h3>
<p>MINORg allows users to specify the identifier of an RPS-BLAST position-specific scoring matrix (PSSM-Id) to further restrict the target sequence to a given domain associated with the PSSM-Id. This could be particularly useful when designing gRNA for genes that do not share conserved domain structures but do share a domain that you wish to knock out.</p>
<section id="local-database">
<h4><span class="section-number">5.4.4.1. </span>Local database<a class="headerlink" href="#local-database" title="Permalink to this headline"></a></h4>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_204_domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">rpsblast</span> <span class="o">=</span> <span class="s2">&quot;/path/to/rpsblast/executable&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="s2">&quot;/path/to/rpsblast/db&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pssm_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;214815&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the above example, gRNA will be generated for the WRKY domain (PSSM-Id 214815 as of CDD database v3.18) of the gene AT5G45050. Users are responsible for providing the PSSM-Id of a domain that exists in the gene. If multiple PSSM-Ids are provided, overlapping domains will be combined and output WILL NOT distinguish between one PSSM-Id or another. Unlike other examples, the database (<a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.db" title="minorg.MINORg.MINORg.db"><code class="xref py py-attr docutils literal notranslate"><span class="pre">db</span></code></a>) is not provided as part of the example files. You will have to download it yourself. See <a class="reference internal" href="parameters.html#rps-blast-local-database"><span class="std std-ref">RPS-BLAST local database</span></a> for more information.</p>
</section>
<section id="remote-database">
<h4><span class="section-number">5.4.4.2. </span>Remote database<a class="headerlink" href="#remote-database" title="Permalink to this headline"></a></h4>
<p>While it is in theory possible to use the remote CDD database &amp; servers instead of local ones, the <code class="docutils literal notranslate"><span class="pre">--remote</span></code> option for the ‘rpsblast’/’rpsblast+’ command from the BLAST+ package has never worked for me. In any case, if your version of local rpsblast is able to access the remote database, you can use <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.remote_rps" title="minorg.MINORg.MINORg.remote_rps"><code class="xref py py-attr docutils literal notranslate"><span class="pre">remote_rps</span></code></a> instead of <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.db" title="minorg.MINORg.MINORg.db"><code class="xref py py-attr docutils literal notranslate"><span class="pre">db</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_204_domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">rpsblast</span> <span class="o">=</span> <span class="s2">&quot;/path/to/rpsblast/executable&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">remote_rps</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pssm_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;214815&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="defining-grna">
<h2><span class="section-number">5.5. </span>Defining gRNA<a class="headerlink" href="#defining-grna" title="Permalink to this headline"></a></h2>
<p>See also: <a class="reference internal" href="parameters.html#pam"><span class="std std-ref">PAM</span></a></p>
<p>By default, MINORg generates 20 bp gRNA using SpCas9’s NGG PAM. You may specify other gRNA length using <a class="reference internal" href="minorg.MINORg.html#id3" title="minorg.MINORg.MINORg.length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">length</span></code></a> and other PAM using <a class="reference internal" href="minorg.MINORg.html#id2" title="minorg.MINORg.MINORg.pam"><code class="xref py py-attr docutils literal notranslate"><span class="pre">pam</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_205_grna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">19</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg</span> <span class="kn">import</span> <span class="n">pam</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pam</span> <span class="o">=</span> <span class="n">pam</span><span class="o">.</span><span class="n">Cas12a</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the example above, MINORg will generate 19 bp gRNA (<code class="docutils literal notranslate"><span class="pre">my_minorg.length</span> <span class="pre">=</span> <span class="pre">19</span></code>) using Cas12a’s unusual 5’ PAM pattern (TTTV&lt;gRNA&gt;) (<code class="docutils literal notranslate"><span class="pre">my_minorg.pam</span> <span class="pre">=</span> <span class="pre">pam.Cas12a</span></code>). MINORg has several built-in PAMs (see <a class="reference internal" href="parameters.html#preset-pam-patterns"><span class="std std-ref">Preset PAM patterns</span></a> for options), and also supports customisable PAM patterns using ambiguous bases and regular expressions (see <a class="reference internal" href="parameters.html#pam"><span class="std std-ref">PAM</span></a> for format). To use preset PAMs, such as in the example above, you will first need to import MINORg’s <a class="reference internal" href="minorg.pam.html#module-minorg.pam"><span class="std std-ref">minorg.pam module</span></a> (<code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">minorg</span> <span class="pre">import</span> <span class="pre">pam</span></code>), then use <code class="docutils literal notranslate"><span class="pre">pam.&lt;preset</span> <span class="pre">pam</span> <span class="pre">alias&gt;</span></code> (such as <code class="docutils literal notranslate"><span class="pre">pam.Cas12a</span></code>) to refer to the desired PAM pattern.</p>
</section>
<section id="filtering-grna">
<h2><span class="section-number">5.6. </span>Filtering gRNA<a class="headerlink" href="#filtering-grna" title="Permalink to this headline"></a></h2>
<p>MINORg supports 3 different gRNA filtering options, all of which can be used together.</p>
<section id="filter-by-gc-content">
<h3><span class="section-number">5.6.1. </span>Filter by GC content<a class="headerlink" href="#filter-by-gc-content" title="Permalink to this headline"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_206_gc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">gc_min</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">gc_max</span> <span class="o">=</span> <span class="mf">0.8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the above example, MINORg will exclude gRNA with less than 20% (<code class="docutils literal notranslate"><span class="pre">my_minorg.gc_min</span> <span class="pre">=</span> <span class="pre">0.2</span></code>) or greater than 80% (<code class="docutils literal notranslate"><span class="pre">my_minorg.gc_min</span> <span class="pre">=</span> <span class="pre">0.8</span></code>) GC content. By default, minimum GC content is 30% and maximum is 70%.</p>
</section>
<section id="filter-by-off-target">
<h3><span class="section-number">5.6.2. </span>Filter by off-target<a class="headerlink" href="#filter-by-off-target" title="Permalink to this headline"></a></h3>
<p>See: <a class="reference internal" href="algorithms.html#off-target-assessment"><span class="std std-ref">Off-target assessment</span></a></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_207_ot_ref&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">screen_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araly2.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araly&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araha1.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_gap</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_mismatch</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the above example, MINORg will screen gRNA for off-targets in:</p>
<ul class="simple">
<li><p>The reference genome (<code class="docutils literal notranslate"><span class="pre">my_minorg.screen_reference</span></code>)</p></li>
<li><p>Four different FASTA files (<code class="docutils literal notranslate"><span class="pre">my_minorg.add_background(&quot;&lt;FASTA&gt;&quot;,</span> <span class="pre">alias</span> <span class="pre">=</span> <span class="pre">&quot;&lt;alias&gt;&quot;)</span></code>)</p>
<ul>
<li><p>The alias keyword argument is optional. If not provided, MINORg will generate a unique alias.</p></li>
<li><p>Note that any AT5G45050 homologues in these four FASTA files will NOT be masked. This means that only gRNA that do not target any AT5G45050 homologues in these four genomes will pass this off-target check.
* To mask homologues in these genomes, you will need to provide a FASTA file containing the sequences of their homologues using <code class="docutils literal notranslate"><span class="pre">my_minorg.mask</span> <span class="pre">=</span> <span class="pre">[&quot;/path/to/to_mask_1.fasta&quot;,</span> <span class="pre">&quot;/path/to/to_mask_2.fasta&quot;]</span></code>. You may use subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a> (see <a class="reference internal" href="#subcommands"><span class="std std-ref">Subcommands</span></a>) to identify these homologues.</p></li>
</ul>
</li>
</ul>
<p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.ot_gap" title="minorg.MINORg.MINORg.ot_gap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ot_gap</span></code></a> and <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.ot_mismatch" title="minorg.MINORg.MINORg.ot_mismatch"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ot_mismatch</span></code></a> control the minimum number of gaps or mismatches off-target gRNA hits must have to be considered non-problematic; any gRNA with at least one problematic gRNA hit will be excluded. By default, both values are set to ‘1’. See <a class="reference internal" href="algorithms.html#off-target-assessment"><span class="std std-ref">Off-target assessment</span></a> for more on the off-target assessment algorithm.</p>
<p>In the case above, <code class="docutils literal notranslate"><span class="pre">my_minorg.screen_reference</span> <span class="pre">=</span> <span class="pre">True</span></code> is actually redundant as the genome from which targets are obtained (which, because of <code class="docutils literal notranslate"><span class="pre">my_minorg.query_reference</span></code>, is the reference genome) are automatically included for background check. However, in the example below, when the targets are from <strong>non-reference genomes</strong>, the reference genome is not automatically included for off-target assessment and thus <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.screen_reference" title="minorg.MINORg.MINORg.screen_reference"><code class="xref py py-attr docutils literal notranslate"><span class="pre">screen_reference</span></code></a> is NOT redundant. Additionally, do note that the genes specified using <code class="xref py py-attr docutils literal notranslate"><span class="pre">gene</span></code> are masked in the reference genome, such that any gRNA hits to them are NOT considered off-target and will NOT be excluded.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_208_ot_nonref&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">screen_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araly2.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araly&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araha1.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_gap</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_mismatch</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<section id="pam-less-off-target-check">
<h4><span class="section-number">5.6.2.1. </span>PAM-less off-target check<a class="headerlink" href="#pam-less-off-target-check" title="Permalink to this headline"></a></h4>
<p>By default, MINORg checks for the presence of PAM sites next to potential off-target hits. If there are no PAM sites, MINORg considers that hit non-problematic. You may override this behaviour by setting <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.ot_pamless" title="minorg.MINORg.MINORg.ot_pamless"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ot_pamless</span></code></a> to <code class="docutils literal notranslate"><span class="pre">True</span></code>. This tells MINORg to mark off-target hits that meet the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.ot_gap" title="minorg.MINORg.MINORg.ot_gap"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ot_gap</span></code></a> or <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.ot_mismatch" title="minorg.MINORg.MINORg.ot_mismatch"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ot_mismatch</span></code></a> thresholds as problematic regardless of the presence or absence of PAM sites nearby.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_209_ot_pamless&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">screen_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araly2.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araly&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araha1.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_gap</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_mismatch</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">ot_pamless</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="skip-off-target-check">
<h4><span class="section-number">5.6.2.2. </span>Skip off-target check<a class="headerlink" href="#skip-off-target-check" title="Permalink to this headline"></a></h4>
<p>To skip off-target check entirely, use <code class="docutils literal notranslate"><span class="pre">background_check</span> <span class="pre">=</span> <span class="pre">False</span></code> when calling <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_210_skipbgcheck&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">background_check</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="filter-by-feature">
<h3><span class="section-number">5.6.3. </span>Filter by feature<a class="headerlink" href="#filter-by-feature" title="Permalink to this headline"></a></h3>
<p>See: <a class="reference internal" href="algorithms.html#within-feature-inference"><span class="std std-ref">Within-feature inference</span></a></p>
<p>By default, when <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.genes" title="minorg.MINORg.MINORg.genes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">genes</span></code></a> is set, MINORg restricts gRNA to coding regions (CDS). For more on how MINORg does this for inferred, unannotated homologues, see <a class="reference internal" href="algorithms.html#within-feature-inference"><span class="std std-ref">Within-feature inference</span></a>. You may change the feature type in which to design gRNA using the attribute <a class="reference internal" href="minorg.MINORg.html#id4" title="minorg.MINORg.MINORg.feature"><code class="xref py py-attr docutils literal notranslate"><span class="pre">feature</span></code></a>. See column 3 of your GFF3 file for valid feature types (see <a class="reference external" href="https://en.wikipedia.org/wiki/General_feature_format">https://en.wikipedia.org/wiki/General_feature_format</a> for more on GFF file format).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_211_withinfeature&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">feature</span> <span class="o">=</span> <span class="s2">&quot;three_prime_UTR&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">background_check</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="generating-minimum-grna-set-s">
<h2><span class="section-number">5.7. </span>Generating minimum gRNA set(s)<a class="headerlink" href="#generating-minimum-grna-set-s" title="Permalink to this headline"></a></h2>
<section id="number-of-sets">
<h3><span class="section-number">5.7.1. </span>Number of sets<a class="headerlink" href="#number-of-sets" title="Permalink to this headline"></a></h3>
<p>By default, MINORg outputs a single gRNA set covering all targets. You may request more (mutually exclusive) sets using the <code class="xref py py-attr docutils literal notranslate"><span class="pre">set</span></code> attribute.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_212_set&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G46260&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46270&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46450&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46470&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46490&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46510&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46520&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">set</span> <span class="o">=</span> <span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="prioritise-non-redundancy">
<h3><span class="section-number">5.7.2. </span>Prioritise non-redundancy<a class="headerlink" href="#prioritise-non-redundancy" title="Permalink to this headline"></a></h3>
<p>By default, MINORg selects gRNA for sets using these criteria in decreasing order of priority:</p>
<ol class="arabic simple">
<li><p>Coverage (of as yet uncovered targets)</p></li>
<li><p>Proximity to 5’ end</p></li>
<li><p>Non-redundancy</p></li>
</ol>
<p>Proximity is only assessed when there is a tie for coverage, and non-redundancy when there is a tie for both coverage and proximity. You may flip the priority of proximity and non-redundancy by setting <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.prioritise_nr" title="minorg.MINORg.MINORg.prioritise_nr"><code class="xref py py-attr docutils literal notranslate"><span class="pre">prioritise_nr</span></code></a> to <code class="docutils literal notranslate"><span class="pre">True</span></code> if you prefer to minimise multiple edits in a single target when using a single set of gRNA. (The priority of coverage is NOT modifiable.)</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_213_nr&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G46260&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46270&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46450&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46470&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46490&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46510&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46520&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">prioritise_nr</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="excluding-grna">
<h3><span class="section-number">5.7.3. </span>Excluding gRNA<a class="headerlink" href="#excluding-grna" title="Permalink to this headline"></a></h3>
<p>You may specify gRNA sequences to exclude from any final gRNA set by providing the path to a FASTA file containing sequences to exclude to <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.exclude" title="minorg.MINORg.MINORg.exclude"><code class="xref py py-attr docutils literal notranslate"><span class="pre">exclude</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_214_exclude&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G46260&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46270&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46450&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46470&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46490&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46510&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46520&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">exclude</span> <span class="o">=</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">sample_exclude_RPS6</span><span class="o">.</span><span class="n">fasta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>The gRNA names in the file passed to <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.exclude" title="minorg.MINORg.MINORg.exclude"><code class="xref py py-attr docutils literal notranslate"><span class="pre">exclude</span></code></a> do not matter. Only the sequences are used when determining whether to exclude a gRNA.</p>
</section>
<section id="accepting-unknown-checks">
<h3><span class="section-number">5.7.4. </span>Accepting unknown checks<a class="headerlink" href="#accepting-unknown-checks" title="Permalink to this headline"></a></h3>
<p>Sometimes, not all filtering checks (GC, background, and feature) are set for all sequences. This is not an issue if you use the full programme (i.e. <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a>), but may be relevant if you are re-generating sets using the ‘minimumset’ subcommand (i.e. <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a>) with a modified mapping file OR a mapping file from the ‘filter’ subcommand where not all filters have been applied.</p>
<p>Let us take a look at ‘sample_custom_check.map’, where we’ve added a custom check called ‘my_custom_check’ in the last column:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gRNA</span> <span class="nb">id</span>       <span class="n">gRNA</span> <span class="n">sequence</span>   <span class="n">target</span> <span class="nb">id</span>       <span class="n">target</span> <span class="n">sense</span>    <span class="n">gRNA</span> <span class="n">strand</span>     <span class="n">start</span>   <span class="n">end</span>     <span class="n">group</span>   <span class="n">background</span>      <span class="n">GC</span>      <span class="n">feature</span> <span class="n">my_custom_check</span>
<span class="n">gRNA_001</span>      <span class="n">CTTCATCTTCTTCTCGAAAT</span>    <span class="n">targetA</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">8</span>       <span class="mi">27</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_001</span>      <span class="n">CTTCATCTTCTTCTCGAAAT</span>    <span class="n">targetB</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">80</span>      <span class="mi">99</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetA</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">37</span>      <span class="mi">56</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">NA</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetB</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">286</span>     <span class="mi">305</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetC</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">109</span>     <span class="mi">128</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">fail</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetD</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">110</span>     <span class="mi">129</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">fail</span>
<span class="n">gRNA_003</span>      <span class="n">ATGTTTTCTTGAGCTTCAGA</span>    <span class="n">targetB</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">38</span>      <span class="mi">57</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">NA</span>
<span class="n">gRNA_003</span>      <span class="n">ATGTTTTCTTGAGCTTCAGA</span>    <span class="n">targetC</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">287</span>     <span class="mi">306</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_003</span>      <span class="n">ATGTTTTCTTGAGCTTCAGA</span>    <span class="n">targetD</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">110</span>     <span class="mi">129</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
</pre></div>
</div>
<p>There are three possible values for check status: ‘pass’, ‘fail’, and ‘NA’.</p>
<p>An invalid/unset check is an ‘NA’. If a check is unset for all entries (as is the case with the check ‘feature’ here), it will be ignored (i.e. the check is treated as ‘pass’ for all entries). However, when a check has been set for some entries but not others (as is the case with the ‘my_custom_check’ check here), MINORg will treat invalid/unset checks as ‘fail’ by default. This is because there isn’t enough information on whether this constitutes a pass or fail for the check, and MINORg prefers to be conservative when outputting gRNA. You may override this behaviour by setting <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.accept_invalid" title="minorg.MINORg.MINORg.accept_invalid"><code class="xref py py-attr docutils literal notranslate"><span class="pre">accept_invalid</span></code></a> to <code class="docutils literal notranslate"><span class="pre">True</span></code>. By doing so, MINORg will treat ‘NA’ as ‘pass’ for all checks.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_215_acceptinvalid&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">parse_grna_map_from_file</span><span class="p">(</span><span class="s2">&quot;/path/to/sample_custom_check.map&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">accept_invalid</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="manually-approve-grna-sets">
<h3><span class="section-number">5.7.5. </span>Manually approve gRNA sets<a class="headerlink" href="#manually-approve-grna-sets" title="Permalink to this headline"></a></h3>
<p>You may opt to manually inspect each gRNA set before MINORg write them to file by using <code class="docutils literal notranslate"><span class="pre">manual</span> <span class="pre">=</span> <span class="pre">True</span></code> when executing <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a> or the minimum set subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_216_manual&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">manual</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

<span class="go">        ID   sequence (Set 1)</span>
<span class="go">        gRNA_001     GGAATACAAGAGATTATCGA</span>
<span class="go">Hit &#39;x&#39; to continue if you are satisfied with these sequences. Otherwise, enter the sequence ID or</span>
<span class="go">sequence of an undesirable gRNA (case-sensitive) and hit the return key to update this list: x</span>

<span class="go">Final gRNA sequence(s) have been written to minorg_gRNA_final.fasta</span>
<span class="go">Final gRNA sequence ID(s), gRNA sequence(s), and target(s) have been written to minorg_gRNA_final.map</span>

<span class="go">1 mutually exclusive gRNA set(s) requested. 1 set(s) found.</span>
<span class="go">Output files have been generated in /path/to/example_216_manual</span>
</pre></div>
</div>
</section>
</section>
<section id="subcommands">
<h2><span class="section-number">5.8. </span>Subcommands<a class="headerlink" href="#subcommands" title="Permalink to this headline"></a></h2>
<p>MINORg comprises of four main steps:</p>
<ol class="arabic simple">
<li><p>Target sequence identification</p></li>
<li><p>Candidate gRNA generation</p></li>
<li><p>gRNA filtering</p></li>
<li><p>Minimum gRNA set generation</p></li>
</ol>
<p>As users may only wish to execute a subset of these steps instead of the full programme (<a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a>), MINORg also provides four subcommands (methods) corresponding to these four steps:</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a></p></li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.grna" title="minorg.MINORg.MINORg.grna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">grna()</span></code></a></p></li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter" title="minorg.MINORg.MINORg.filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a>, which itself calls three other methods</p>
<ul class="simple">
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_background" title="minorg.MINORg.MINORg.filter_background"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_background()</span></code></a></p></li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_feature" title="minorg.MINORg.MINORg.filter_feature"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_feature()</span></code></a></p></li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_gc" title="minorg.MINORg.MINORg.filter_gc"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_gc()</span></code></a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a></p></li>
</ol>
<p>The subcommands may be useful if you already have a preferred off-target/on-target assessment software. In this case, you may execute subcommands <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a> and <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.grna" title="minorg.MINORg.MINORg.grna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">grna()</span></code></a>, submit the gRNA output by MINORg for off-target/on-target assessment, update the .map file output by MINORg with the status of each gRNA for that off-target/on-target assessment, and execute <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a> to obtain a desired number of minimum gRNA sets. Note that if you do this, you should re-read the updated .map file into MINORg using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.parse_grna_map_from_file" title="minorg.MINORg.MINORg.parse_grna_map_from_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">parse_grna_map_from_file()</span></code></a> so MINORg can replace the gRNA data stored in memory with your updated gRNA data.</p>
<p>Each subcommand may require a different combination of attributes.</p>
<section id="subcommand-seq">
<h3><span class="section-number">5.8.1. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a><a class="headerlink" href="#subcommand-seq" title="Permalink to this headline"></a></h3>
<p>The <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a> subcommand identifies target sequences, whether by extracting them from a reference genome or inferring homologues in unannotated genomes. All parameters introduced in <a class="reference internal" href="#defining-target-sequences"><span class="std std-ref">Defining target sequences</span></a> (except attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code>) and <a class="reference internal" href="#defining-reference-genomes"><span class="std std-ref">Defining reference genomes</span></a> apply. If you already have a FASTA file containing your target sequences, you may set <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code> to the path of that FASTA file and skip this subcommand.</p>
<p>This step will output target sequences into a file ending with ‘_targets.fasta’. This filename will be stored at attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_217_subcmdseq&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">extend_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/sample_gene.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT1G10920&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">seq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">target</span>
<span class="go">&#39;/path/to/example_217_subcmdseq/minorg/minorg_gene_targets.fasta&#39;</span>
</pre></div>
</div>
</section>
<section id="subcommand-grna">
<h3><span class="section-number">5.8.2. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.grna" title="minorg.MINORg.MINORg.grna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">grna()</span></code></a><a class="headerlink" href="#subcommand-grna" title="Permalink to this headline"></a></h3>
<p>The <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.grna" title="minorg.MINORg.MINORg.grna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">grna()</span></code></a> subcommand generates gRNA within target sequences from a target file. Unlike the command line version, it <strong>DOES NOT</strong> incorporate parts of the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a> and <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter" title="minorg.MINORg.MINORg.filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a> subcommands. All parameters introduced in <a class="reference internal" href="#defining-grna"><span class="std std-ref">Defining gRNA</span></a> apply.</p>
<p>By default, .map and FASTA files of gRNA sequences will be written to files. You may override this behaviour by setting <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.auto_update_files" title="minorg.MINORg.MINORg.auto_update_files"><code class="xref py py-attr docutils literal notranslate"><span class="pre">auto_update_files</span></code></a> to <code class="docutils literal notranslate"><span class="pre">False</span></code> or using <code class="docutils literal notranslate"><span class="pre">auto_update_files</span> <span class="pre">=</span> <span class="pre">False</span></code> when instantiating a <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a> object (e.g. <code class="docutils literal notranslate"><span class="pre">my_minorg(directory</span> <span class="pre">=</span> <span class="pre">&quot;/path/to/output/dir&quot;,</span> <span class="pre">auto_update_files</span> <span class="pre">=</span> <span class="pre">False)</span></code>). In this case, only the FASTA file will be written. To manually write files, you should use the following methods. If you do not supply an output file path, it will be automatically generated:</p>
<ul class="simple">
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_all_grna_map" title="minorg.MINORg.MINORg.write_all_grna_map"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_all_grna_map()</span></code></a>: write .map file containing all candidate gRNA (no checks will be set by <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.grna" title="minorg.MINORg.MINORg.grna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">grna()</span></code></a> so all entries in check fields will be ‘NA’)</p>
<ul>
<li><p>Path to output file will be stored at <code class="xref py py-attr docutils literal notranslate"><span class="pre">grna_map</span></code></p></li>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_all.map</p></li>
</ul>
</li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_all_grna_fasta" title="minorg.MINORg.MINORg.write_all_grna_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_all_grna_fasta()</span></code></a>: write FASTA file containing all candidate gRNA</p>
<ul>
<li><p>Path to output file will be stored at <code class="xref py py-attr docutils literal notranslate"><span class="pre">grna_fasta</span></code></p></li>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_all.fasta</p></li>
</ul>
</li>
</ul>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_218_subcmdgrna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna</span><span class="p">()</span> <span class="c1">## default 3&#39; NGG PAM</span>
<span class="go">PAM pattern: .{20}(?=[GATC]GG)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna_hits</span>
<span class="go">gRNAHits(gRNA = 201)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg</span> <span class="kn">import</span> <span class="n">pam</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pam</span> <span class="o">=</span> <span class="n">pam</span><span class="o">.</span><span class="n">Cas12a</span> <span class="c1">## 5&#39; TTTV PAM</span>
<span class="go">PAM pattern: (?&lt;=TTT[ACG]).{20}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna_hits</span>
<span class="go">gRNAHits(gRNA = 95)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pam</span> <span class="o">=</span> <span class="s2">&quot;ATV.&quot;</span>
<span class="go">PAM pattern: (?&lt;=AT[ACG]).{20}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna_hits</span>
<span class="go">gRNAHits(gRNA = 267)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_all_grna_fasta</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna_fasta</span>
<span class="go">&#39;/path/to/example_218_subcmdgrna/minorg/minorg_gRNA_all.fasta&#39;</span>
</pre></div>
</div>
<p>gRNA data is stored at the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">grna_hits</span></code>, and it prints the number of gRNA as a string representation. In the above example, 201 different gRNA are generated from the target sequences in the target file “sample_CDS.fasta”. We then decided we want to generate gRNA for Cas12a instead, which has a 5’ TTTV PAM pattern. This yields us 95 different gRNA. Finally we decided to try a completely made up 5’ ATV PAM pattern, netting us 267 different gRNA in the end. Satisfied, we wrote the sequences of these gRNA to file, and printed the path of the file.</p>
</section>
<section id="subcommand-filter">
<h3><span class="section-number">5.8.3. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter" title="minorg.MINORg.MINORg.filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a><a class="headerlink" href="#subcommand-filter" title="Permalink to this headline"></a></h3>
<p>The <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter" title="minorg.MINORg.MINORg.filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a> subcommand takes in a compulsory MINORg .map file (which can be read using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.parse_grna_map_from_file" title="minorg.MINORg.MINORg.parse_grna_map_from_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">parse_grna_map_from_file()</span></code></a>) and rewrites some/all checks. You can execute all filters (GC, off-target, and feature) using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter" title="minorg.MINORg.MINORg.filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a>, or execute checks separately using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_gc" title="minorg.MINORg.MINORg.filter_gc"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_gc()</span></code></a>, <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_background" title="minorg.MINORg.MINORg.filter_background"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_background()</span></code></a>, and <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_feature" title="minorg.MINORg.MINORg.filter_feature"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_feature()</span></code></a>.</p>
<p>By default, gRNA sequences and map files will be updated automatically whenever any of the filtering methods is called. You may override this behaviour by setting <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.auto_update_files" title="minorg.MINORg.MINORg.auto_update_files"><code class="xref py py-attr docutils literal notranslate"><span class="pre">auto_update_files</span></code></a> to <code class="docutils literal notranslate"><span class="pre">False</span></code> or using <code class="docutils literal notranslate"><span class="pre">auto_update_files</span> <span class="pre">=</span> <span class="pre">False</span></code> when instantiating a <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a> object (e.g. <code class="docutils literal notranslate"><span class="pre">my_minorg(directory</span> <span class="pre">=</span> <span class="pre">&quot;/path/to/output/dir&quot;,</span> <span class="pre">auto_update_files</span> <span class="pre">=</span> <span class="pre">False)</span></code>). To manually write files, you should use the following methods. If you do not supply an output file path, it will be automatically generated:</p>
<ul class="simple">
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_all_grna_map" title="minorg.MINORg.MINORg.write_all_grna_map"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_all_grna_map()</span></code></a>: write .map file containing all candidate gRNA and checks</p>
<ul>
<li><p>Path to output file will be stored at <code class="xref py py-attr docutils literal notranslate"><span class="pre">grna_map</span></code></p></li>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_all.map</p></li>
</ul>
</li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_all_grna_fasta" title="minorg.MINORg.MINORg.write_all_grna_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_all_grna_fasta()</span></code></a>: write FASTA file containing all candidate gRNA</p>
<ul>
<li><p>Path to output file will be stored at <code class="xref py py-attr docutils literal notranslate"><span class="pre">grna_fasta</span></code></p></li>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_all.fasta</p></li>
<li><p>This file will NOT be auto updated as it is not affected by filtering check status</p></li>
</ul>
</li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_pass_grna_map" title="minorg.MINORg.MINORg.write_pass_grna_map"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_pass_grna_map()</span></code></a>: write .map file containing all passing gRNA</p>
<ul>
<li><p>Path to output file will be stored at <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.pass_map" title="minorg.MINORg.MINORg.pass_map"><code class="xref py py-attr docutils literal notranslate"><span class="pre">pass_map</span></code></a></p></li>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_pass.map</p></li>
</ul>
</li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_pass_grna_fasta" title="minorg.MINORg.MINORg.write_pass_grna_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_pass_grna_fasta()</span></code></a>: write FASTA file containing all passing gRNA</p>
<ul>
<li><p>Path to output file will be stored at <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.pass_fasta" title="minorg.MINORg.MINORg.pass_fasta"><code class="xref py py-attr docutils literal notranslate"><span class="pre">pass_fasta</span></code></a></p></li>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_pass.fasta</p></li>
</ul>
</li>
</ul>
<p>In all cases, you may rename the gRNA using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.rename_grna" title="minorg.MINORg.MINORg.rename_grna"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rename_grna()</span></code></a>, which takes in the path of a FASTA file that contains the gRNA sequences you wish to rename with sequence IDs of the names you wish to rename them to. This method should be used before you call any of the above methods to write gRNA to file.</p>
<section id="subcommand-filter-gc">
<h4><span class="section-number">5.8.3.1. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_gc" title="minorg.MINORg.MINORg.filter_gc"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_gc()</span></code></a><a class="headerlink" href="#subcommand-filter-gc" title="Permalink to this headline"></a></h4>
<p>All parameters introduced in <a class="reference internal" href="#filter-by-gc-content"><span class="std std-ref">Filter by GC content</span></a> apply.</p>
<section id="filtering-by-gc-content-after-calling-full">
<h5><span class="section-number">5.8.3.1.1. </span>Filtering by GC content after calling <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a><a class="headerlink" href="#filtering-by-gc-content-after-calling-full" title="Permalink to this headline"></a></h5>
<p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_gc" title="minorg.MINORg.MINORg.filter_gc"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_gc()</span></code></a> can be used on an active MINORg object even if you’ve already called <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_219_subcmdfilter_gc&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">&lt;set up parameters&gt;</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">gc_min</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">gc_max</span> <span class="o">=</span> <span class="mf">0.8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_gc</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="filtering-gc-content-on-output-of-another-minorg-run">
<h5><span class="section-number">5.8.3.1.2. </span>Filtering GC content on output of another MINORg run<a class="headerlink" href="#filtering-gc-content-on-output-of-another-minorg-run" title="Permalink to this headline"></a></h5>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_219_subcmdfilter_gc_pt2&quot;</span><span class="p">,</span> <span class="n">auto_update_files</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">parse_grna_map_from_file</span><span class="p">(</span><span class="s2">&quot;/path/to/sample_custom_check.map&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">gc_min</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">gc_max</span> <span class="o">=</span> <span class="mf">0.8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_gc</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_pass_grna_fasta</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="subcommand-filter-background">
<h4><span class="section-number">5.8.3.2. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_background" title="minorg.MINORg.MINORg.filter_background"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_background()</span></code></a><a class="headerlink" href="#subcommand-filter-background" title="Permalink to this headline"></a></h4>
<p>All parameters introduced in <a class="reference internal" href="#filter-by-off-target"><span class="std std-ref">Filter by off-target</span></a> apply. Additionally, you should supply target sequences to <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code> so that MINORg can mask them (this tells MINORg that any gRNA hits to them is in fact on-target and NOT off-target). Any additional sequences to be masked may be provided to <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.mask" title="minorg.MINORg.MINORg.mask"><code class="xref py py-attr docutils literal notranslate"><span class="pre">mask</span></code></a> as a list of paths to FASTA files. If you have set <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.screen_reference" title="minorg.MINORg.MINORg.screen_reference"><code class="xref py py-attr docutils literal notranslate"><span class="pre">screen_reference</span></code></a> to <code class="docutils literal notranslate"><span class="pre">True</span></code> to include reference genome(s) (see <a class="reference internal" href="#multiple-reference-genomes"><span class="std std-ref">Multiple reference genomes</span></a> for how to specify multiple reference genomes) in the off-target screen, you may specify a FASTA file of sequences of genes to be masked to <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.mask" title="minorg.MINORg.MINORg.mask"><code class="xref py py-attr docutils literal notranslate"><span class="pre">mask</span></code></a> as well. You can generate these sequences using the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.seq" title="minorg.MINORg.MINORg.seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">seq()</span></code></a> subcommand, but <strong>MAKE SURE TO USE A DIFFERENT MINORg OBJECT AND DIRECTORY TO AVOID OVERWRITING ANY PREVIOUSLY GENERATED FILES</strong>.</p>
<section id="filtering-background-after-calling-full">
<h5><span class="section-number">5.8.3.2.1. </span>Filtering background after calling <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a><a class="headerlink" href="#filtering-background-after-calling-full" title="Permalink to this headline"></a></h5>
<p>Let us first execute MINORg.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_220_subcmdfilter_bg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G46450&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46470&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46490&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46510&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46520&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">set</span> <span class="o">=</span> <span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">background_check</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>In the code above, we skipped off-target check using <code class="docutils literal notranslate"><span class="pre">background_check</span> <span class="pre">=</span> <span class="pre">False</span></code> when executing <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a>. But we’ve changed out mind and would like to screen the reference genome and the non-reference genomes that these targets are from AND we don’t want our gRNA to be able to target any genes in ‘subset_9944.fasta’ and ‘subset_9947’. We also want to tell MINORg that it’s okay if a gRNA has off-target effects in homologous genes AT5G46260 and AT5G46270 in the reference genome. We can do that using the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter" title="minorg.MINORg.MINORg.filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter()</span></code></a> subcommand, followed by the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a> subcommand to regenerate minimum sets.</p>
<p>In order to do all this, we will have to get the gene sequences of AT5G46260 and AT5G46270 in order to mask them in the reference genome. We can do this using the <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.get_reference_seq" title="minorg.MINORg.MINORg.get_reference_seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_reference_seq()</span></code></a> method.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ot_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_220_subcmdfilter_bg_tomask&quot;</span><span class="p">)</span> <span class="c1">## different directory</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ot_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ot_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G46260&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46270&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fout_to_mask</span> <span class="o">=</span> <span class="n">ot_minorg</span><span class="o">.</span><span class="n">mkfname</span><span class="p">(</span><span class="s2">&quot;ref_to_mask.fasta&quot;</span><span class="p">)</span> <span class="c1">## MINORg has a built-in method to generate file names within the output directory</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ot_minorg</span><span class="o">.</span><span class="n">get_reference_seq</span><span class="p">(</span><span class="n">fout</span> <span class="o">=</span> <span class="n">fout_to_mask</span><span class="p">)</span> <span class="c1">## this method will return a dictionary of sequences, but will also write to file if &#39;fout&#39; is used</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ot_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>Now that we have the reference sequences to mask, we can pass the file name to <code class="docutils literal notranslate"><span class="pre">my_minorg</span></code>‘s <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.mask" title="minorg.MINORg.MINORg.mask"><code class="xref py py-attr docutils literal notranslate"><span class="pre">mask</span></code></a> attribute, add our background files using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_background" title="minorg.MINORg.MINORg.add_background"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_background()</span></code></a>, set <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.screen_reference" title="minorg.MINORg.MINORg.screen_reference"><code class="xref py py-attr docutils literal notranslate"><span class="pre">screen_reference</span></code></a> to <code class="docutils literal notranslate"><span class="pre">True</span></code>, call <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_background" title="minorg.MINORg.MINORg.filter_background"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_background()</span></code></a> to update off-target checks for all candidate gRNA, and execute <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a> to regenerate our minimum gRNA sets. You may also wish to call <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_all_grna_map" title="minorg.MINORg.MINORg.write_all_grna_map"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_all_grna_map()</span></code></a>, <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_pass_grna_map" title="minorg.MINORg.MINORg.write_pass_grna_map"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_pass_grna_map()</span></code></a>, and/or <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.write_pass_grna_fasta" title="minorg.MINORg.MINORg.write_pass_grna_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_pass_grna_fasta()</span></code></a> to update the gRNA FASTA and .map files if <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.auto_update_files" title="minorg.MINORg.MINORg.auto_update_files"><code class="xref py py-attr docutils literal notranslate"><span class="pre">auto_update_files</span></code></a> has been set to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">mask</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fout_to_mask</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9944.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9944&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9947.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9947&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">screen_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_background</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="filtering-background-on-output-of-another-minorg-run">
<h5><span class="section-number">5.8.3.2.2. </span>Filtering background on output of another MINORg run<a class="headerlink" href="#filtering-background-on-output-of-another-minorg-run" title="Permalink to this headline"></a></h5>
<p>Alternatively, if the orginal <code class="docutils literal notranslate"><span class="pre">my_minorg</span></code> object no longer exists, whether because you’ve closed the IDE session or deleted the object, you can read its .map file into a new <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a> object using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.parse_grna_map_from_file" title="minorg.MINORg.MINORg.parse_grna_map_from_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">parse_grna_map_from_file()</span></code></a> like below. In this case, you can pass the IDs of the additional genes to be masked together with the original genes to <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.genes" title="minorg.MINORg.MINORg.genes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">genes</span></code></a> and don’t need to use <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.get_reference_seq" title="minorg.MINORg.MINORg.get_reference_seq"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_reference_seq()</span></code></a>. Since we’re no longer querying ‘subset_9654.fasta’ and ‘subset_9655.fasta’, we can use <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_background" title="minorg.MINORg.MINORg.add_background"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_background()</span></code></a> to tell MINORg to search for off-target effects in them. And don’t forget to also provide the FASTA file of target sequences to <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code> so MINORg can mask them!:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_220_subcmdfilter_bg_new&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">parse_grna_map_from_file</span><span class="p">(</span><span class="s2">&quot;/path/to/example_220_subcmdfilter_bg/minorg/minorg_gRNA_all.map&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_220_subcmdfilter_bg/minorg/minorg_gene_targets.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G46260&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46270&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46450&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46470&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46490&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46510&quot;</span><span class="p">,</span> <span class="s2">&quot;AT5G46520&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9944.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9944&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9947.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9947&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">screen_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">filter_background</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="subcommand-filter-feature">
<h4><span class="section-number">5.8.3.3. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_feature" title="minorg.MINORg.MINORg.filter_feature"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_feature()</span></code></a><a class="headerlink" href="#subcommand-filter-feature" title="Permalink to this headline"></a></h4>
<p>All parameters introduced in <a class="reference internal" href="#filter-by-feature"><span class="std std-ref">Filter by feature</span></a> apply. Additionally, you will need to provide a FASTA file of target sequences (attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code>), reference genome(s) (see <a class="reference internal" href="#defining-reference-genomes"><span class="std std-ref">Defining reference genomes</span></a>), and genes (attribute <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.genes" title="minorg.MINORg.MINORg.genes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">genes</span></code></a>). The specified reference gene(s) will be extracted from the reference genome(s) and aligned with target sequence(s) in order for MINORg to infer feature boundaries in target sequence(s). See <a class="reference internal" href="algorithms.html#within-feature-inference"><span class="std std-ref">Within-feature inference</span></a> for the algorithm of how feature boundaries are inferred.</p>
<section id="filtering-feature-after-calling-full">
<h5><span class="section-number">5.8.3.3.1. </span>Filtering feature after calling <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a><a class="headerlink" href="#filtering-feature-after-calling-full" title="Permalink to this headline"></a></h5>
<p>Let us first execute MINORg.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_221_subcmdfilter_feature&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
</pre></div>
</div>
<p>By default, MINORg sets the desired feature to ‘CDS’. You can re-assess and overwrite the ‘feature’ check in the .map file to only allow gRNA in the 3’ UTR using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.filter_feature" title="minorg.MINORg.MINORg.filter_feature"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_feature()</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">feature</span> <span class="o">=</span> <span class="s2">&quot;three_prime_UTR&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_feature</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="filtering-feature-on-output-of-another-minorg-run">
<h5><span class="section-number">5.8.3.3.2. </span>Filtering feature on output of another MINORg run<a class="headerlink" href="#filtering-feature-on-output-of-another-minorg-run" title="Permalink to this headline"></a></h5>
<p>As with <a class="reference internal" href="#filtering-background-on-output-of-another-minorg-run"><span class="std std-ref">Filtering background on output of another MINORg run</span></a>, we can read in the output of a previous MINORg execution and filter that. This requires the .map file ending with ‘_all.map’ (parse using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.parse_grna_map_from_file" title="minorg.MINORg.MINORg.parse_grna_map_from_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">parse_grna_map_from_file()</span></code></a>) as well as a FASTA file of target sequences (specify using <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code>).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_221_subcmdfilter_feature_new&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">parse_grna_map_from_file</span><span class="p">(</span><span class="s2">&quot;/path/to/example_221_subcmdfilter_feature/minorg/minorg_gRNA_all.map&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_221_subcmdfilter_feature/minorg/minorg_gene_targets.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G45050&quot;</span><span class="p">]</span> <span class="c1">## MINORg needs to know which reference genes to align to targets to in order to infer feature ranges</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">feature</span> <span class="o">=</span> <span class="s2">&quot;three_prime_UTR&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">filter_feature</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="subcommand-minimumset">
<h3><span class="section-number">5.8.4. </span>Subcommand <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a><a class="headerlink" href="#subcommand-minimumset" title="Permalink to this headline"></a></h3>
<p>The <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a> subcommand generates mutually exclusive minimum set(s) of gRNA, where each set is capable of covering all targets. It requires a MINORg .map file (the one that ends in ‘_gRNA_pass.map’ is sufficient, but ‘_gRNA_all.map’ would allow for filtering by a custom combination of fields). All parameters introduced in <a class="reference internal" href="#generating-minimum-grna-set-s"><span class="std std-ref">Generating minimum gRNA set(s)</span></a> apply.</p>
<p>This step will write final gRNA sequences into a file ending with ‘_gRNA_final.fasta’. A file ending with ‘_gRNA_final.map’ that maps gRNA to their targets will also be generated. You may optionally specify the location of the FASTA and .map output files using:</p>
<ul class="simple">
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.final_map" title="minorg.MINORg.MINORg.final_map"><code class="xref py py-attr docutils literal notranslate"><span class="pre">final_map</span></code></a>: path of .map file containing gRNA in final set(s)</p>
<ul>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_final.map</p></li>
</ul>
</li>
<li><p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.final_fasta" title="minorg.MINORg.MINORg.final_fasta"><code class="xref py py-attr docutils literal notranslate"><span class="pre">final_fasta</span></code></a>: path of FASTA file containing gRNA in final set(s)</p>
<ul>
<li><p>If output file is not specified, the output file will be &lt;output_directory&gt;/&lt;prefix&gt;/&lt;prefix&gt;_gRNA_final.fasta</p></li>
</ul>
</li>
</ul>
<section id="regenerating-minimum-sets-after-calling-full">
<h4><span class="section-number">5.8.4.1. </span>Regenerating minimum sets after calling <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.full" title="minorg.MINORg.MINORg.full"><code class="xref py py-meth docutils literal notranslate"><span class="pre">full()</span></code></a><a class="headerlink" href="#regenerating-minimum-sets-after-calling-full" title="Permalink to this headline"></a></h4>
<p><a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.minimumset" title="minorg.MINORg.MINORg.minimumset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimumset()</span></code></a> can also be used on an active MINORg object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_222_subcmdminimumset_pt2&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">&lt;set up parameters&gt;</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">set</span> <span class="o">=</span> <span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">()</span> <span class="c1">## regenerate up to 5 gRNA sets</span>
</pre></div>
</div>
</section>
<section id="generating-minimum-sets-from-output-of-another-minorg-run">
<h4><span class="section-number">5.8.4.2. </span>Generating minimum sets from output of another MINORg run<a class="headerlink" href="#generating-minimum-sets-from-output-of-another-minorg-run" title="Permalink to this headline"></a></h4>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_222_subcmdminimumset&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">parse_grna_map_from_file</span><span class="p">(</span><span class="s2">&quot;/path/to/example_203_query/minorg/minorg_gRNA_all.map&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_203_query/minorg/minorg_gene_targets.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">prioritise_nr</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">set</span> <span class="o">=</span> <span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">(</span><span class="n">gc_check</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In order for MINORg to better assess a gRNA’s proximity to the 5’ end (of hopefully sense strand) of a target in the event a tie-breaker is necessary, it is strongly suggested that target sequences be provided to <code class="xref py py-attr docutils literal notranslate"><span class="pre">target</span></code> so MINORg knows how long a target sequence is. This is especially so if the target sequences are antisense ones (you can check this using the .map file) generated by MINORg’s inferences of homologues in unannotated genomes. In the example above, we’ve asked MINORg to ignore the GC content check when generating minimum sets (<code class="docutils literal notranslate"><span class="pre">my_minorg.minimumset(gc_check</span> <span class="pre">=</span> <span class="pre">False)</span></code>).</p>
</section>
</section>
<section id="chaining-subcommands">
<h3><span class="section-number">5.8.5. </span>Chaining subcommands<a class="headerlink" href="#chaining-subcommands" title="Permalink to this headline"></a></h3>
<p>You may use subcommands separately if you’d like to inspect the outcome of each step and/or repeat a step with different parameters before proceeding with the next. MINORg tracks the output of previous steps, so you do not need to read them into MINORg before executing the next step.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_223_subcmd&quot;</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">,</span> <span class="n">thread</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araly2.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_Araly2.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araly2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT5G66900&quot;</span><span class="p">,</span> <span class="s2">&quot;AL8G44500.v2.1&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">seq</span><span class="p">()</span> <span class="c1">## generate target sequences</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">target</span> <span class="c1">## print path to FASTA file containing target sequences</span>
<span class="go">&#39;/path/to/example_223_subcmd/minorg/minorg_gene_targets.fasta&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna</span><span class="p">()</span>
<span class="go">PAM pattern: .{20}(?=[GATC]GG)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">screen_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_background</span><span class="p">()</span>
<span class="go">Masking on-targets</span>
<span class="go">Finding off-targets</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">(</span><span class="s2">&quot;background&quot;</span><span class="p">)</span>
<span class="go">gRNAHits(gRNA = 395)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_background</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araha1.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;araha1&quot;</span><span class="p">)</span> <span class="c1">## add background file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_background</span><span class="p">()</span> <span class="c1">## repeat background check with additional background file</span>
<span class="go">Masking on-targets</span>
<span class="go">Finding off-targets</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">(</span><span class="s2">&quot;background&quot;</span><span class="p">)</span> <span class="c1">## updated set of passing gRNA</span>
<span class="go">gRNAHits(gRNA = 250)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_gc</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">(</span><span class="s2">&quot;GC&quot;</span><span class="p">)</span>
<span class="go">gRNAHits(gRNA = 355)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">(</span><span class="s2">&quot;background&quot;</span><span class="p">,</span> <span class="s2">&quot;GC&quot;</span><span class="p">)</span>
<span class="go">gRNAHits(gRNA = 223)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">()</span> <span class="c1">## gRNA filtered for all valid checks (at this point, background and GC)</span>
<span class="go">/path/to/minorg/grna.py:823: MINORgWarning: The following hit checks have not been set: feature</span>
<span class="go">gRNAHits(gRNA = 223)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_feature</span><span class="p">()</span> <span class="c1">## by default, MINORg only retains gRNA in CDS</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">(</span><span class="s2">&quot;feature&quot;</span><span class="p">)</span>
<span class="go">gRNAHits(gRNA = 324)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">valid_grna</span><span class="p">()</span>
<span class="go">gRNAHits(gRNA = 181)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">minimumset</span><span class="p">(</span><span class="n">manual</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

<span class="go">        ID   sequence (Set 1)</span>
<span class="go">        gRNA_026     GTCGTTTCCGGAGACTATGA</span>
<span class="go">Hit &#39;x&#39; to continue if you are satisfied with these sequences. Otherwise, enter the sequence ID or</span>
<span class="go">sequence of an undesirable gRNA (case-sensitive) and hit the return key to update this list: gRNA_026</span>

<span class="go">        ID   sequence (Set 1)</span>
<span class="go">        gRNA_223     TCAATCTCCATCATAGTCTC</span>
<span class="go">Hit &#39;x&#39; to continue if you are satisfied with these sequences. Otherwise, enter the sequence ID or</span>
<span class="go">sequence of an undesirable gRNA (case-sensitive) and hit the return key to update this list: x</span>

<span class="go">Final gRNA sequence(s) have been written to /path/to/example_223_subcmd/minorg/minorg_gRNA_final.fasta</span>
<span class="go">Final gRNA sequence ID(s), gRNA sequence(s), and target(s) have been written to</span>
<span class="go">/path/to/example_223_subcmd/minorg/minorg_gRNA_final.map</span>

<span class="go">1 mutually exclusive gRNA set(s) requested. 1 set(s) found.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_all_grna_map</span><span class="p">()</span> <span class="c1">## write .map file containing check information for all candidate gRNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_all_grna_fasta</span><span class="p">()</span> <span class="c1">## write FASTA file containing all candidate gRNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_pass_grna_map</span><span class="p">()</span> <span class="c1">## write .map file containing information for valid gRNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_pass_grna_fasta</span><span class="p">()</span> <span class="c1">## write FASTA file containing valid gRNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span> <span class="c1">## remove temporary files</span>
</pre></div>
</div>
<p>It is highly recommended that you execute <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.resolve" title="minorg.MINORg.MINORg.resolve"><code class="xref py py-meth docutils literal notranslate"><span class="pre">resolve()</span></code></a> to remove any temporary files generated.</p>
</section>
</section>
<section id="defining-reference-genomes">
<h2><span class="section-number">5.9. </span>Defining reference genomes<a class="headerlink" href="#defining-reference-genomes" title="Permalink to this headline"></a></h2>
<section id="single-reference-genome">
<h3><span class="section-number">5.9.1. </span>Single reference genome<a class="headerlink" href="#single-reference-genome" title="Permalink to this headline"></a></h3>
<p>See example in <a class="reference internal" href="#reference-gene-s-as-targets"><span class="std std-ref">Reference gene(s) as targets</span></a>.</p>
</section>
<section id="multiple-reference-genomes">
<h3><span class="section-number">5.9.2. </span>Multiple reference genomes<a class="headerlink" href="#multiple-reference-genomes" title="Permalink to this headline"></a></h3>
<p>See also: <a class="reference internal" href="parameters.html#reference"><span class="std std-ref">Reference</span></a></p>
<p>You may specify genes from multiple reference genomes so long as those reference genomes have also been added using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_224_multiref&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_TAIR10.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;TAIR10&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araly2.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_Araly2.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;Araly2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_Araha1.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_Araha1.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;Araha1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT1G33560&quot;</span><span class="p">,</span> <span class="s2">&quot;AL1G47950.v2.1&quot;</span><span class="p">,</span> <span class="s2">&quot;Araha.3012s0003.v1.1&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the example above, MINORg will design gRNA for 3 highly conserved paralogues in 3 different species. Note that you should be careful that any gene IDs you use should either be unique across all reference genomes OR be shared only among your target genes. Otherwise, MINORg will treat any undesired genes with the same gene IDs as targets as well.</p>
</section>
<section id="non-standard-reference">
<h3><span class="section-number">5.9.3. </span>Non-standard reference<a class="headerlink" href="#non-standard-reference" title="Permalink to this headline"></a></h3>
<section id="non-standard-genetic-code">
<h4><span class="section-number">5.9.3.1. </span>Non-standard genetic code<a class="headerlink" href="#non-standard-genetic-code" title="Permalink to this headline"></a></h4>
<p>When using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.pssm_ids" title="minorg.MINORg.MINORg.pssm_ids"><code class="xref py py-attr docutils literal notranslate"><span class="pre">pssm_ids</span></code></a>, users should ensure that the correct genetic code has been specified for reference genomes using the <code class="docutils literal notranslate"><span class="pre">genetic_code</span></code> keyword argument when adding reference genomes using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a>, as MINORg has to first translate CDS into peptides for domain search using RPS-BLAST. The default genetic code is the Standard Code. Please refer to <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a> for genetic code numbers and names.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_225_geneticcode&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_yeast_mt.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_yeast_mt.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;yeast_mt&quot;</span><span class="p">,</span> <span class="n">genetic_code</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span> <span class="c1">## specify genetic code here</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene-Q0275&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pssm_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;366140&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>In the above example, the gene ‘gene-Q0275’ is a yeast mitochondrial gene, and <code class="docutils literal notranslate"><span class="pre">my_minorg.pssm_ids</span> <span class="pre">=</span> <span class="pre">[&quot;366140&quot;]</span></code> specifies the PSSM-Id for the COX3 domain in the Cdd v3.18 RPS-BLAST database. The genetic code number for yeast mitochondrial code is ‘3’.</p>
<p>As a failsafe, MINORg does not terminate translated peptide sequences at the first stop codon. This ensures that any codons after an incorrectly translated premature stop codon will still be translated. Typically, a handful of mistranslated codons can still result in the correct RPS-BLAST domain hits, although hit scores may be slightly lower. Nevertheless, to ensure maximum accuracy, the correct genetic code is preferred.</p>
</section>
<section id="non-standard-gff3-attribute-field-names">
<h4><span class="section-number">5.9.3.2. </span>Non-standard GFF3 attribute field names<a class="headerlink" href="#non-standard-gff3-attribute-field-names" title="Permalink to this headline"></a></h4>
<p>See also: <a class="reference internal" href="parameters.html#attribute-modification"><span class="std std-ref">Attribute modification</span></a></p>
<p>MINORg requires standard attribute field names in GFF3 files in order to properly map subfeatures to their parent features (e.g. map CDS to mRNA, and mRNA to gene). Non-standard field names should be mapped to standard ones using the <code class="docutils literal notranslate"><span class="pre">attr_mod</span></code> (for ‘attribute modification’) keyword argument when adding reference genomes using <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.add_reference" title="minorg.MINORg.MINORg.add_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_reference()</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_226_attrmod&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_ref_irgsp.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/subset_ref_irgsp.gff&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;irgsp&quot;</span><span class="p">,</span> <span class="n">attr_mod</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;mRNA&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="s2">&quot;Locus_id&quot;</span><span class="p">}})</span> <span class="c1">## specify attribute modifications</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Os01t0100100&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
<p>The IRGSP 1.0 reference genome for rice (<em>Oryza sativa</em> subsp. Nipponbare) uses a non-standard attribute field name for mRNA entries in their GFF3 file. Instead of ‘Parent’, which is the standard name of the field used to map a feature to its parent feature, mRNA entries in the IRGSP 1.0 annotation use ‘Locus_id’. See <a class="reference internal" href="parameters.html#attribute-modification"><span class="std std-ref">Attribute modification</span></a> for more details on how to format the input to <code class="docutils literal notranslate"><span class="pre">attr_mod</span></code>.</p>
</section>
</section>
</section>
<section id="multithreading">
<h2><span class="section-number">5.10. </span>Multithreading<a class="headerlink" href="#multithreading" title="Permalink to this headline"></a></h2>
<p>MINORg supports multi-threading in order to process files in parallel. Any excess threads may also be used for BLAST. This is most useful when you are querying multiple genomes, have multiple reference genomes, or multiple background sequences.</p>
<p>To run MINORg with parallel processing, set <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg.thread" title="minorg.MINORg.MINORg.thread"><code class="xref py py-attr docutils literal notranslate"><span class="pre">thread</span></code></a> to the desired number of threads.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;/path/to/example_227_thread&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">extend_reference</span><span class="p">(</span><span class="s2">&quot;/path/to/sample_gene.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/sample_CDS.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AT1G10920&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9654.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9654&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s2">&quot;/path/to/subset_9655.fasta&quot;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s2">&quot;9655&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">thread</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">full</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="differences-between-cli-and-python-versions">
<h2><span class="section-number">5.11. </span>Differences between CLI and Python versions<a class="headerlink" href="#differences-between-cli-and-python-versions" title="Permalink to this headline"></a></h2>
<p>Note that, unlike the command line, the Python package does not support aliases even if the config file has been set up appropriately for command line executions. Therefore, there are no true equivalents to <code class="docutils literal notranslate"><span class="pre">--cluster</span></code>, <code class="docutils literal notranslate"><span class="pre">--indv</span></code>, or <code class="docutils literal notranslate"><span class="pre">--reference</span></code>.</p>
<section id="to-specify-cluster-genes">
<h3><span class="section-number">5.11.1. </span>To specify cluster genes<a class="headerlink" href="#to-specify-cluster-genes" title="Permalink to this headline"></a></h3>
<p>Analogous to <code class="docutils literal notranslate"><span class="pre">--cluster</span></code> and <code class="docutils literal notranslate"><span class="pre">--gene</span></code>.</p>
<p><strong>Correct:</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AT5G46260&#39;</span><span class="p">,</span><span class="s1">&#39;AT5G46270&#39;</span><span class="p">,</span><span class="s1">&#39;AT5G46450&#39;</span><span class="p">,</span><span class="s1">&#39;AT5G46470&#39;</span><span class="p">,</span><span class="s1">&#39;AT5G46490&#39;</span><span class="p">,</span><span class="s1">&#39;AT5G46510&#39;</span><span class="p">,</span><span class="s1">&#39;AT5G46520&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p><strong>Incorrect:</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">cluster_set</span> <span class="o">=</span> <span class="s1">&#39;/path/to/subset_cluster_mapping.txt&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">cluster</span> <span class="o">=</span> <span class="s1">&#39;RPS6&#39;</span>
</pre></div>
</div>
<p>Attributes ‘cluster_set’ and ‘cluster’ do not exist. This does not throw error now but will cause problems later.</p>
</section>
<section id="to-specify-query-fasta-files">
<h3><span class="section-number">5.11.2. </span>To specify query FASTA files<a class="headerlink" href="#to-specify-query-fasta-files" title="Permalink to this headline"></a></h3>
<p>Analogous to <code class="docutils literal notranslate"><span class="pre">--indv</span></code> and <code class="docutils literal notranslate"><span class="pre">--query</span></code>.</p>
<p><strong>Correct:</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s1">&#39;/path/to/subset_9654.fasta&#39;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s1">&#39;9654&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_query</span><span class="p">(</span><span class="s1">&#39;/path/to/subset_9655.fasta&#39;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s1">&#39;9655&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Incorrect:</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genome_set</span> <span class="o">=</span> <span class="s1">&#39;/path/to/subset_genome_mapping.txt&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">indv</span> <span class="o">=</span> <span class="s1">&#39;9654,9655&#39;</span>
</pre></div>
</div>
<p>Attributes ‘genome_set’ and ‘indv’ do not exist. This does not throw error now but will cause problems later.</p>
</section>
<section id="to-specify-reference-genomes">
<h3><span class="section-number">5.11.3. </span>To specify reference genomes<a class="headerlink" href="#to-specify-reference-genomes" title="Permalink to this headline"></a></h3>
<p>Analogous to <code class="docutils literal notranslate"><span class="pre">--reference</span></code>, <code class="docutils literal notranslate"><span class="pre">--assembly</span></code>, <code class="docutils literal notranslate"><span class="pre">--annotation</span></code>, <code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code>, and <code class="docutils literal notranslate"><span class="pre">--genetic-code</span></code>.</p>
<p><strong>Correct:</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s1">&#39;/path/to/TAIR10.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/TARI10.gff3&#39;</span><span class="p">,</span> <span class="n">alias</span> <span class="o">=</span> <span class="s1">&#39;TAIR10&#39;</span><span class="p">,</span> <span class="n">genetic_code</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">atr_mod</span> <span class="o">=</span> <span class="p">{})</span>
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">attr_mod</span></code> and <code class="docutils literal notranslate"><span class="pre">genetic_code</span></code> are optional if the annotation uses standard attribute field names and the standard genetic code, which the example above does.</p>
<p><strong>Incorrect:</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">reference_set</span> <span class="o">=</span> <span class="s1">&#39;/path/to/arabidopsis_genomes.txt&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="s1">&#39;TAIR10&#39;</span>
<span class="go">AttributeError: can&#39;t set attribute</span>
</pre></div>
</div>
<p>Attributes ‘reference_set’ does not exist, and ‘reference’ is a property that users are not allowed to directly modify.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial_cli.html" class="btn btn-neutral float-left" title="4. Tutorial (Command line)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="parameters.html" class="btn btn-neutral float-right" title="6. Parameters" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>